---
title: "MethylKit"
author: "Laura H Spencer"
date: "11/17/2022"
output: html_document
knit: (function(input, ...) {
    rmarkdown::render(
      input,
      output_file = paste0(
        xfun::sans_ext(input), '-', Sys.Date(), '.html'
      ),
      envir = globalenv()
    )
  })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Load libraries

```{r message=FALSE, warning=FALSE, results=FALSE}
list.of.packages <- c("tidyverse", "reshape2", "here", "methylKit", "ggforce", "matrixStats", "Pstat", "viridis", "plotly", "readr", "GenomicRanges", "vegan", "factoextra", "PerformanceAnalytics", "corrplot", "janitor", "googlesheets4", "ggrepel", "clipr", "plotly", "wordcloud", "tm", "slam", "eulerr", "forcats", "ggpubr") #add new libraries here 

new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)

# Load all libraries 
lapply(list.of.packages, FUN = function(X) {
  do.call("require", list(X)) 
})

# Load MethylKit
if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

BiocManager::install("methylKit")
require(methylKit)

`%!in%` = Negate(`%in%`)
```

### Obtain session information

```{r}
sessionInfo()
```

### Import bismark alignment .bams into MethylKit. NOTE: This was done on Sedna! 

I ran this code interactively on Sedna 
`srun --pty /bin/bash` # First, I entered an interactive Sedna node
`module load R` # Then, load R
`R` #start R interactively 
Now proceed with the following code 

```{r}
# ## AGAIN- this code was run interactively on Sedna! 
# # Install and load programs
# 
# # Load MethylKit
# if (!require("BiocManager", quietly = TRUE))
#   install.packages("BiocManager")
# 
# BiocManager::install("methylKit")
# 
# require(methylKit)
# require(tidyverse)
# 
# # Load sample info dataframe 
# load(file="/home/lspencer/DuMOAR/R/sample.info")
# 
# # Triple check that treatments are in the order of sequence sample number (1-20)
# (sample.info %>% arrange(sample_seq))$sample # List out library prep sample number 
# (sample.info %>% arrange(sample_seq))$sample_seq # List out sequence sample 
# (sample.info %>% arrange(sample_seq))$high_or_low_co2 # List out treatment
# 
# # Read .bam files into Methylkit on Sedna 
# (file.list=list("/scratch/lspencer/DuMOAR/aligned/trim-before-concat/CH01_06_S1_R1_bismark_bt2_pe.deduplicated.sorted.bam",
#                "/scratch/lspencer/DuMOAR/aligned/trim-before-concat/CH01_14_S2_R1_bismark_bt2_pe.deduplicated.sorted.bam",
#                "/scratch/lspencer/DuMOAR/aligned/trim-before-concat/CH01_22_S3_R1_bismark_bt2_pe.deduplicated.sorted.bam",
#                "/scratch/lspencer/DuMOAR/aligned/trim-before-concat/CH01_38_S4_R1_bismark_bt2_pe.deduplicated.sorted.bam",
#                "/scratch/lspencer/DuMOAR/aligned/trim-before-concat/CH03_04_S5_R1_bismark_bt2_pe.deduplicated.sorted.bam",
#                "/scratch/lspencer/DuMOAR/aligned/trim-before-concat/CH03_15_S6_R1_bismark_bt2_pe.deduplicated.sorted.bam",
#                "/scratch/lspencer/DuMOAR/aligned/trim-before-concat/CH03_33_S7_R1_bismark_bt2_pe.deduplicated.sorted.bam",
#                "/scratch/lspencer/DuMOAR/aligned/trim-before-concat/CH05_01_S8_R1_bismark_bt2_pe.deduplicated.sorted.bam",
#                "/scratch/lspencer/DuMOAR/aligned/trim-before-concat/CH05_06_S9_R1_bismark_bt2_pe.deduplicated.sorted.bam",
#                "/scratch/lspencer/DuMOAR/aligned/trim-before-concat/CH05_21_S10_R1_bismark_bt2_pe.deduplicated.sorted.bam",
#                "/scratch/lspencer/DuMOAR/aligned/trim-before-concat/CH05_24_S11_R1_bismark_bt2_pe.deduplicated.sorted.bam",
#                "/scratch/lspencer/DuMOAR/aligned/trim-before-concat/CH07_06_S12_R1_bismark_bt2_pe.deduplicated.sorted.bam",
#                "/scratch/lspencer/DuMOAR/aligned/trim-before-concat/CH07_11_S13_R1_bismark_bt2_pe.deduplicated.sorted.bam",
#                "/scratch/lspencer/DuMOAR/aligned/trim-before-concat/CH07_24_S14_R1_bismark_bt2_pe.deduplicated.sorted.bam",
#                "/scratch/lspencer/DuMOAR/aligned/trim-before-concat/CH09_02_S15_R1_bismark_bt2_pe.deduplicated.sorted.bam",
#                "/scratch/lspencer/DuMOAR/aligned/trim-before-concat/CH09_13_S16_R1_bismark_bt2_pe.deduplicated.sorted.bam",
#                "/scratch/lspencer/DuMOAR/aligned/trim-before-concat/CH09_28_S17_R1_bismark_bt2_pe.deduplicated.sorted.bam",
#                "/scratch/lspencer/DuMOAR/aligned/trim-before-concat/CH10_01_S18_R1_bismark_bt2_pe.deduplicated.sorted.bam",
#                "/scratch/lspencer/DuMOAR/aligned/trim-before-concat/CH10_08_S19_R1_bismark_bt2_pe.deduplicated.sorted.bam",
#                "/scratch/lspencer/DuMOAR/aligned/trim-before-concat/CH10_11_S20_R1_bismark_bt2_pe.deduplicated.sorted.bam"))
# 
# #### The following command reads sorted BAM files and calls methylation percentage per base, and creates a methylRaw object for CpG methylation. It also assigns minimum coverage of 2x and the treatments (in this case, the CO2 treatment)
# 
# meth_obj = processBismarkAln(location = file.list, sample.id = list("1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18", "19", "20"), 
#                           assembly = "PGA_assembly.fasta", read.context="CpG", mincov=2, 
#                           treatment = as.numeric((sample.info %>% arrange(sample_seq))$high_or_low_co2))
# 
# #### Save the MethylKit object; re-doing the previous step is memory/time intensive, so best to use the saved object moving forward. 
# 
# save(meth_obj, file = "/home/lspencer/DuMOAR/R/meth_obj")
```

```{r}
getwd()
```

#### Read in Bismark summary stats and some library prep stats 

```{r}
# Read in data from GoogleSheet 
sample.info <- read_sheet("https://docs.google.com/spreadsheets/d/1ym0XnYVts98tIUCn0kIaU6VuvqxzV7LoSx9RHwLdiIs/edit?usp=sharing", "DNAisolations_only") %>% 
  clean_names() %>% 
  mutate_at(vars(tank, treatment, high_or_low_co2, developmental_stage, dna_siolation_batch), as.factor) %>%
  
  # Join with bismark summary report 
  right_join(

  # Read in bismark summary report
    read_delim(file="../results/bismark/scaffold-only/bismark_summary_report.txt", delim="\t") %>% mutate(File = str_replace_all(File, '_R1_bismark_bt2_pe.bam', "")) %>%
  clean_names() %>% 
   separate(file, into = c("a", "b", "sample_seq"), remove = F, sep = "_") %>%
   mutate(sample_prep=paste(a, b, sep = "-")) %>%
    mutate(sample_seq=as.numeric(str_replace_all(sample_seq, "S", ""))) %>%
   dplyr::select(file, sample_prep, sample_seq, everything(), -a, -b) %>%
  mutate(perc_totalread_unique=unique_reads_remaining/total_reads, 
         CpGs_Meth=100*methylated_cp_gs/(methylated_cp_gs+unmethylated_cp_gs), 
         CHGs_Meth=100*methylated_chgs/(methylated_chgs+unmethylated_chgs), 
         CHHs_Meth=100*methylated_ch_hs/(methylated_ch_hs+unmethylated_ch_hs)),

  by = c("sample"="sample_prep")) %>%

# add mbd library prep stats
  left_join(
  
  read_sheet("https://docs.google.com/spreadsheets/d/1ym0XnYVts98tIUCn0kIaU6VuvqxzV7LoSx9RHwLdiIs/edit?usp=sharing", "MBD_calcs") %>% 
  clean_names() %>% 
  dplyr::select(sample, total_recovery_ng, percent_recovery_percent, library_bioanlyzer_mean_fragment_size_bp,
         library_bioanalyzer_molarity_pmol_l, qubit_concentration_ng_u_l), 

  by = "sample")


save(sample.info, file="../data/sample.info")
load(file="../data/sample.info")
```

### Investigate possible factors influencing low %CpGmeth 

#### Correlation bubble plot 

```{r}
cor(sample.info %>% dplyr::select_if(is.numeric)) %>% corrplot(tl.cex=0.65) 
```

#### Look closer at factors that appear to correlate with CpGs_Meth to get correlation statistics  

```{r, warning=FALSE}
#trace("chart.Correlation", edit=T) #to edit function
# In line 17, change the cex value for what you want, i changed it to 2

chart.Correlation(sample.info %>% 
                    dplyr::select(CpGs_Meth, a260_280, mean_mw_highest_peak, aligned_reads, 
                           unaligned_reads, ambiguously_aligned_reads, no_genomic_sequence, 
                           unique_reads_remaining, total_cs, perc_totalread_unique, CHGs_Meth, CHHs_Meth), 
                  histogram=F, pch=19) 
```

Question asked by Zymo (creater of PicoMethyl kit)  - any correlation between the samples with abnormal CpG methylation and samples that had more yield from the MBD enrichment step?

Answer: no not really 

```{r}
chart.Correlation(sample.info %>% 
                    dplyr::select(CpGs_Meth, total_recovery_ng, percent_recovery_percent, 
                           library_bioanlyzer_mean_fragment_size_bp, library_bioanalyzer_molarity_pmol_l, 
                           qubit_concentration_ng_u_l), 
                  histogram=F, pch=19) 
# CpG methylation is not associated with MBD yield


# Use this code to generate 
# sample.info %>% dplyr::select(sample_seq, high_or_low_co2, CpGs_Meth, 
#                        total_recovery_ng, percent_recovery_percent, library_bioanlyzer_mean_fragment_size_bp, 
#                        library_bioanalyzer_molarity_pmol_l, qubit_concentration_ng_u_l) %>%
#   pivot_longer(cols = c("total_recovery_ng", "percent_recovery_percent", "library_bioanlyzer_mean_fragment_size_bp", 
#                        "library_bioanalyzer_molarity_pmol_l", "qubit_concentration_ng_u_l"), names_to = "MBD.stat") %>%
#   mutate(MBD.stat=as.factor(MBD.stat)) %>%
# 
# ggplot(aes(x=value, y=CpGs_Meth)) + 
#   geom_point(aes(label=sample_seq, color=high_or_low_co2), size=3) + 
#   scale_color_manual(values=c("red", "blue")) +
#            geom_text(aes(label=sample_seq, color=high_or_low_co2), vjust = -0.7) + 
#   theme_minimal() + geom_smooth(method = "lm", se=T, color="black", linetype="dotted", fill="gray80") +
# ggtitle("% CpG Methylation ~ MBD Stats") + 
#   facet_wrap(~MBD.stat, scales = "free")

# Weak correlation with MBD library fragment size, but not convincing
ggplot(sample.info, aes(x=library_bioanlyzer_mean_fragment_size_bp, y=CpGs_Meth)) +
  geom_point(aes(label=sample, color=high_or_low_co2), size=3) +
  scale_color_manual(values=c("red", "blue")) +
           geom_text(aes(label=sample, color=high_or_low_co2), vjust = -0.7) +
  theme_minimal() + geom_smooth(method = "lm", se=T, color="black", linetype="dotted", fill="gray80") +
    annotate("text", x = 500, y = 17, label = "r = 0.42", size=6) + 
  ggtitle("% CpG Methylation ~ library_bioanlyzer_mean_fragment_size_bp") 

# Low CpG methylation is associated with high CHH methylation 
ggplot(sample.info, aes(x=CHHs_Meth, y=CpGs_Meth)) +
  geom_point(aes(label=sample, color=high_or_low_co2), size=3) +
  scale_color_manual(values=c("red", "blue")) +
           geom_text(aes(label=sample, color=high_or_low_co2), vjust = -0.7) +
  theme_minimal() + geom_smooth(method = "lm", se=T, color="black", linetype="dotted", fill="gray80") +
    annotate("text", x = 33, y = 70, label = "r = -0.59", size=6) + 
  ggtitle("% CpG Methylation ~ CHHs_Meth") 
```

Correlations between CpGs_Meth and the following:  
- a260_280: 0.52* - meh     
- mean_mw_highest_peak: 0.51* - not predictive, but low CpG meth samples have lower mean MW.    
- aligned_reads: 0.66** - low CpG meth = fewer aligned reads    
- unaligned_reads: -0.77***   
- ambiguously_aligned_reads: -0.60**   
- unique_reads_remaining: 0.73***  
- total_cs: 0.72***  
- perc_totalread_unique: 0.82***  
- CHGs_Meth: -0.38  
- CHHs_Meth: -0.59**  

```{r, warning=FALSE,message=FALSE}
ggplot(sample.info, aes(x=a260_280, y=CpGs_Meth)) + 
  geom_point(aes(label=sample, color=high_or_low_co2), size=3) + 
  scale_color_manual(values=c("red", "blue")) +
           geom_text(aes(label=sample, color=high_or_low_co2), vjust = -0.7) + 
  theme_minimal() + geom_smooth(method = "lm", se=T, color="black", linetype="dotted", fill="gray80") +
  annotate("text", x = 1.5, y = 75, label = "r = 0.52", size=6) + 
ggtitle("% CpG Methylation ~ a260_280")
```

```{r, warning=FALSE,message=FALSE}
ggplot(sample.info, aes(x=mean_mw_highest_peak, y=CpGs_Meth)) + 
  geom_point(aes(label=sample, color=high_or_low_co2), size=3) + 
  scale_color_manual(values=c("red", "blue")) +
           geom_text(aes(label=sample, color=high_or_low_co2), vjust = -0.7) + 
  theme_minimal() + geom_smooth(method = "lm", se=T, color="black", linetype="dotted", fill="gray80") +
  annotate("text", x = 10000, y = 100, label = "r = 0.51", size=6) + 
ggtitle("% CpG Methylation ~ Mean MW (highest peak)")
```

```{r, warning=FALSE,message=FALSE}
ggplot(sample.info, aes(x=aligned_reads, y=CpGs_Meth)) + 
  geom_point(aes(label=sample, color=high_or_low_co2), size=3) + 
  scale_color_manual(values=c("red", "blue")) +
           geom_text(aes(label=sample, color=high_or_low_co2), vjust = -0.7) + 
  theme_minimal() + geom_smooth(method = "lm", se=T, color="black", linetype="dotted", fill="gray80") +
  annotate("text", x = 6e6, y = 85, label = "r = 0.66", size=6) + 
ggtitle("% CpG Methylation ~ # of aligned reads")
```
```{r, warning=FALSE,message=FALSE}
ggplot(sample.info, aes(x=unaligned_reads, y=CpGs_Meth)) + 
  geom_point(aes(label=sample, color=high_or_low_co2), size=3) + 
  scale_color_manual(values=c("red", "blue")) +
           geom_text(aes(label=sample, color=high_or_low_co2), vjust = -0.7) + 
  theme_minimal() + geom_smooth(method = "lm", se=T, color="black", linetype="dotted", fill="gray80") +
  annotate("text", x = 1e7, y = 85, label = "r = -0.77", size=6) + 
ggtitle("% CpG Methylation ~ # unaligned reads")
```

```{r, warning=FALSE,message=FALSE}
ggplot(sample.info, aes(x=ambiguously_aligned_reads, y=CpGs_Meth)) + 
  geom_point(aes(label=sample, color=high_or_low_co2), size=3) + 
  scale_color_manual(values=c("red", "blue")) +
           geom_text(aes(label=sample, color=high_or_low_co2), vjust = -0.7) + 
  theme_minimal() + geom_smooth(method = "lm", se=T, color="black", linetype="dotted", fill="gray80") +
  annotate("text", x = 5e6, y = 85, label = "r = -0.60", size=6) + 
ggtitle("% CpG Methylation ~ # ambiguously aligned reads")
```
```{r, warning=FALSE,message=FALSE}
ggplot(sample.info, aes(x=no_genomic_sequence, y=CpGs_Meth)) + 
  geom_point(aes(label=sample, color=high_or_low_co2), size=3) + 
  scale_color_manual(values=c("red", "blue")) +
           geom_text(aes(label=sample, color=high_or_low_co2), vjust = -0.7) + 
  theme_minimal() + geom_smooth(method = "lm", se=T, color="black", linetype="dotted", fill="gray80") +
  #annotate("text", x = 285, y = 75, label = "r = -0.93", size=6) + 
ggtitle("% CpG Methylation ~ # unaligned- no genomic sequence")
```


No. of reads not aligning due to no genomic sequence corrlates very well with % methylation, but the actual # of reads is super low across all samples (<300!). What does this mean?! 


```{r, warning=FALSE,message=FALSE}
ggplot(sample.info, aes(x=unique_reads_remaining, y=CpGs_Meth)) + 
  geom_point(aes(label=sample, color=high_or_low_co2), size=3) + 
  scale_color_manual(values=c("red", "blue")) +
           geom_text(aes(label=sample, color=high_or_low_co2), vjust = -0.7) + 
  theme_minimal() + geom_smooth(method = "lm", se=T, color="black", linetype="dotted", fill="gray80") +
  annotate("text", x = 3.5e6, y = 95, label = "r = 0.73", size=6) + 
ggtitle("% CpG Methylation ~ # unique aligned reads")
```

```{r, warning=FALSE,message=FALSE}
ggplot(sample.info, aes(x=total_cs, y=CpGs_Meth)) + 
  geom_point(aes(label=sample, color=high_or_low_co2), size=3) + 
  scale_color_manual(values=c("red", "blue")) +
           geom_text(aes(label=sample, color=high_or_low_co2), vjust = -0.7) + 
  theme_minimal() + geom_smooth(method = "lm", se=T, color="black", linetype="dotted", fill="gray80") +
  annotate("text", x = 6e7, y = 90, label = "r = 0.72", size=6) + 
ggtitle("% CpG Methylation ~ total # Cs")
```
```{r, warning=FALSE,message=FALSE}
ggplot(sample.info, aes(x=perc_totalread_unique, y=CpGs_Meth)) + 
  geom_point(aes(label=sample, color=high_or_low_co2), size=3) + 
  scale_color_manual(values=c("red", "blue")) +
           geom_text(aes(label=sample, color=high_or_low_co2), vjust = -0.7) + 
  theme_minimal() + geom_smooth(method = "lm", se=T, color="black", linetype="dotted", fill="gray80") +
  annotate("text", x = .15, y = 85, label = "r = 0.82", size=6) + 
ggtitle("% CpG Methylation ~ % total reads uniquely aligned")
```

```{r, warning=FALSE,message=FALSE}
ggplot(sample.info, aes(x=CHGs_Meth, y=CpGs_Meth)) + 
  geom_point(aes(label=sample, color=high_or_low_co2), size=3) + 
  scale_color_manual(values=c("red", "blue")) +
           geom_text(aes(label=sample, color=high_or_low_co2), vjust = -0.7) + 
  theme_minimal() + geom_smooth(method = "lm", se=T, color="black", linetype="dotted", fill="gray80") +
  annotate("text", x = 2.25, y = 70, label = "r = -0.38", size=6) + 
ggtitle("% CpG Methylation ~ % CHG Methylation")
```

```{r, warning=FALSE,message=FALSE}
ggplot(sample.info, aes(x=CHHs_Meth, y=CpGs_Meth)) + 
  geom_point(aes(label=sample, color=high_or_low_co2), size=3) + 
  scale_color_manual(values=c("red", "blue")) +
           geom_text(aes(label=sample, color=high_or_low_co2), vjust = -0.7) + 
  theme_minimal() + geom_smooth(method = "lm", se=T, color="black", linetype="dotted", fill="gray80") +
  annotate("text", x = 30, y = 75, label = "r = -0.59", size=6) + 
ggtitle("% CpG Methylation ~ % CHH Methylation")
```


### Load methylKit object that was created on Sedna 

```{r}
load("../results/methylkit/meth_obj")
```

#### What does the resulting methylkit object look like? 

```{r}
# Check out data for sample #1 
head(meth_obj[[1]])
```

### Check the basic stats about the methylation data - coverage and percent methylation.

#### First look at % CpG methylation for each sampe 

```{r}
for (i in 1:20) {
getMethylationStats(meth_obj[[i]],plot=T,both.strands=TRUE)
     mtext(paste("Sample", i, sep=" "), side=3, line = -6)
}
```
#### Now look at coverage for each sample 

```{r}
for (i in 1:20) {
 getCoverageStats(meth_obj[[i]],plot=TRUE,both.strands=TRUE) 
   mtext(paste("Sample", i, sep=" "), side=3, line = -6)
}
```

### Generate CpG methylation histograms after filtering for coverage

#### Filter out loci where coverage is less than 5x or greater than 100x. 

```{r}
meth_5x=filterByCoverage(meth_obj,lo.count=5,lo.perc=NULL,
                                      hi.count=100,hi.perc=NULL)
save(meth_5x, file="../results/methylkit/meth_5x")
load(file="../results/methylkit/meth_5x")
```

#### Now look at CpG methylation after filtering for 5x coverage 

```{r}
for (i in 1:20) {
getMethylationStats(meth_5x[[i]],plot=T,both.strands=TRUE)
  mtext(paste("Sample", i, sep=" "), side=3, line = -6)
}
```

#### Filter out loci where coverage is less than 10x or greater than 100x. 

```{r}
meth_10x=filterByCoverage(meth_obj,lo.count=10,lo.perc=NULL,
                                      hi.count=100,hi.perc=NULL)
save(meth_10x, file="../results/methylkit/meth_10x")
load(file="../results/methylkit/meth_10x")
```

#### Now look at CpG methylation after filtering for 10x coverage 

```{r}
for (i in 1:20) {
getMethylationStats(meth_10x[[i]],plot=T,both.strands=TRUE)
  mtext(paste("Sample", i, sep=" "), side=3, line = -6)
}
```

### Unite methylation data into one object for methylkit functions

#### column bind all samples, and destrand

```{r}
meth_unite=methylKit::unite(meth_obj, destrand=TRUE, min.per.group=1L)
save(meth_unite, file = "../results/methylkit/meth_unite") #save object to file
load(file = "../results/methylkit/meth_unite") #load object from file if needed
```

### Cluster analysis 

```{r}
clusterSamples(meth_unite, dist="correlation", method="ward", plot=TRUE)
```

```{r}
PCASamples(meth_unite, screeplot = T)
PCASamples(meth_unite)
```

#### Reshape methylation to inspect % methylation at various filtering thresholds 
 
```{r}
# Here - use all methylation data
#load(file = "../results/methylkit/meth_unite") #save object to file

# NOTE: since the sample IDs in the meth_unite() object are sequential from 1:20, the sample IDs in the column names containing data (coverage, numCs, numTs) are correct in this instance. 

meth_unite_reshaped <- meth_unite %>% 
  melt(id=c("chr", "start", "end", "strand"), value.name = "count") %>%
  drop_na(count) %>%
  mutate(stat=gsub('[0-9]+', '', variable)) %>%
  mutate(sample=gsub('[a-zA-Z]+', '', variable)) %>%
  dplyr::select(-variable) %>%
  pivot_wider(names_from = stat, values_from = count) %>%
  mutate(percMeth = 100*(numCs/coverage)) %>%
  mutate(sample=as.numeric(sample))

save(meth_unite_reshaped, file = "../results/methylkit/meth_unite_reshaped")
load(file = "../results/methylkit/meth_unite_reshaped")  
```

```{r, cache=TRUE, eval=FALSE}
# # Here - use a smaller data set, already filtered for min coverage 5x 
# meth_unite_5x <- methylKit::unite(meth_5x, destrand=TRUE, min.per.group=1L)
# save(meth_unite_5x, file = "../results/methylkit/meth_unite_5x") #save object to file
# load(file = "../results/methylkit/meth_unite_5x") #save object to file
# 
# meth_unite_5x_reshaped <- meth_unite_5x %>%
#   melt(id=c("chr", "start", "end", "strand"), value.name = "count") %>%
#   drop_na(count) %>%
#   mutate(stat=gsub('[0-9]+', '', variable)) %>%
#   mutate(sample=gsub('[a-zA-Z]+', '', variable)) %>%
#   dplyr::select(-variable) %>%
#   pivot_wider(names_from = stat, values_from = count) %>%
#   mutate(percMeth = 100*(numCs/coverage)) %>%
#   mutate(sample=as.numeric(sample))
# 
# save(meth_unite_5x_reshaped, file = "meth_unite_5x_reshaped")
# save(meth_unite_5x_reshaped, file = "../results/methylkit/meth_unite_5x_reshaped")
```

#### No coverage threshold, what is mean % methylated and how many loci are there?

```{r, warning=F}
meth_unite_reshaped %>% 
  group_by(sample) %>%
  summarize(mean = mean(percMeth), nloci = n()) %>%
  left_join(sample.info[c("treatment", "sample", "sample_seq", "high_or_low_co2")] %>% rename("sample"="sample_prep"), by=c("sample"="sample_seq")) %>%
  ggplot(aes(x=nloci, y=mean, label=sample_prep)) + geom_point(aes(colour=high_or_low_co2)) + 
  geom_label_repel(label.size=0.05, min.segment.length=0.25,aes(colour=high_or_low_co2)) + theme_minimal() +
  ggtitle("% methylation ~ # loci, no coverage minimum") +
  scale_color_manual(values = c("red", "blue")) + 
  theme(legend.position = "bottom") +
  geom_smooth(method = "loess", se=T, color="black", linetype="dotted", fill="gray80")
```

#### 5x coverage, what is mean % methylated and how many loci are there?

```{r, warning=F}
meth_unite_reshaped %>% 
  group_by(sample) %>%
  filter(coverage>=5) %>%
  summarize(mean = mean(percMeth), nloci = n()) %>%
  left_join(sample.info[c("treatment", "sample", "sample_seq", "high_or_low_co2")] %>% rename("sample"="sample_prep"), 
            by=c("sample"="sample_seq")) %>%
  ggplot(aes(x=nloci, y=mean, label=sample_prep)) + 
  geom_point(aes(colour=high_or_low_co2)) + 
  geom_label_repel(label.size=0.05, min.segment.length=0.15,aes(colour=high_or_low_co2)) + theme_minimal() +
  ggtitle("% methylation ~ # loci, 5x coverage minimum") +
  scale_color_manual(values = c("red", "blue")) + 
  theme(legend.position = "bottom") +
  geom_smooth(method = "loess", se=T, color="black", linetype="dotted", fill="gray80")
```

#### 10x coverage, what is mean % methylated and how many loci are there?

```{r, warning=F}
meth_unite_reshaped %>% 
  group_by(sample) %>%
  filter(coverage>=10) %>%
  summarize(mean = mean(percMeth), nloci = n()) %>% 
  left_join(sample.info[c("treatment", "sample", "sample_seq", "high_or_low_co2")] %>% rename("sample"="sample_prep"), 
            by=c("sample"="sample_seq")) %>%
  ggplot(aes(x=nloci, y=mean, label=sample_prep)) + 
  geom_point(aes(colour=high_or_low_co2)) + 
  geom_label_repel(label.size=0.05, min.segment.length=0.15,aes(colour=high_or_low_co2)) + theme_minimal() +
  ggtitle("% methylation ~ # loci, 10x coverage minimum") + xlab("# Loci") + ylab("Mean % CpG Methylation") +
  scale_color_manual(values = c("red", "blue")) + 
  geom_smooth(method = "loess", se=T, color="black", linetype="dotted", fill="gray80")
```

#### What is mean % methylated with 15x coverage?

NOTE: label = "sample_prep" has the format "CH##-##", while "sample" uses the numbers 1-20. 

```{r, warning=F}
sample.info %>% dplyr::select(sample, sample_seq) %>% arrange(sample_seq)

meth_unite_reshaped %>% 
  group_by(sample) %>%
  filter(coverage>=15) %>%
  summarize(mean = mean(percMeth), nloci = n()) %>%
  left_join(sample.info[c("treatment", "sample", "sample_seq", "high_or_low_co2")] %>% rename("sample"="sample_prep"), 
            by=c("sample"="sample_seq")) %>%
  ggplot(aes(x=nloci, y=mean, label=sample_prep)) + 
  geom_point(aes(colour=high_or_low_co2)) + 
  geom_label_repel(label.size=0.03, min.segment.length=0.15,aes(colour=high_or_low_co2)) + theme_minimal() +
  ggtitle("% methylation ~ # loci, 15x coverage minimum") +
  scale_color_manual(values = c("red", "blue")) + 
  geom_smooth(method = "loess", se=T, color="black", linetype="dotted", fill="gray80")
```
#### Is % methylation a function of coverage within a sample? 

The following samples have low CpG methylation: 5, 6, 8 (kinda), 9, 13, 14, 16, 19. Check out each sample individually 

FYI here are sample-prep vs. sample_seq labels, weird samples have *

CH01-06	= 1			
CH01-14	= 2			
CH01-22	= 3			
CH01-38	= 4			
CH03-04	= 5*			
CH03-15	= 6*			
CH03-33	= 7			
CH05-01	= 8*			
CH05-06	= 9*			
CH05-21	= 10	
CH05-24	= 11			
CH07-06	= 12			
CH07-11	= 13*			
CH07-24	= 14*			
CH09-02	= 15			
CH09-13	= 16*			
CH09-28	= 17			
CH10-01	= 18			
CH10-08	= 19*			
CH10-11	= 20	

Weird samples have lots of loci with 0% methylation 

```{r}
meth_unite_reshaped %>%
  filter(coverage>=5) %>%
  ggplot() + geom_point(aes(x=coverage, y=percMeth), size=0.05) +
  theme_minimal() + 
  facet_wrap(~sample)
```

**Bad samples:** all have mean % methylation <70%
control pco2: 5, 6, 13, 14
high pco2: 9, 16, 19, 8 (kinda)

**Good samples:** all have mean % methylation >= 70% and nloci meeting 10x > 3e5
control pCO2: 1, 2, 3, 4, 7, 12
high pCO2:    10, 11, 15, 17, 18, 20

At 10x coverage, here's the mean % methylation and number of loci per sample 

```{r}
meth_unite_reshaped %>% 
  group_by(sample) %>%
  filter(coverage>=10) %>%
  #filter(sample %in%  c("1", "2", "3", "4", "7", "10", "11", "12", "15", "17", "18", "20")) %>%
  #filter(sample %!in%  c("1", "2", "3", "4", "7", "10", "11", "12", "15", "17", "18", "20")) %>%
  summarize(mean = mean(percMeth), nloci = n()) %>% 
  arrange(sample) #%>% dplyr::select(nloci) %>% write_clip()
```


```{r}
# Summary stats by "good" and "bad" samples  

# Bad samples
sample.info %>% filter(sample_seq %!in%  c("1", "2", "3", "4", "7", "10", "11", "12", "15", "17", "18", "20")) %>%
  summarise(mean.aligned=mean(perc_totalread_unique), sd.aligned=sd(perc_totalread_unique),
            min.aligned=min(perc_totalread_unique), max.aligned=max(perc_totalread_unique),
            mean.reads=mean(total_reads), sd.reads=sd(total_reads),
            mean.CpGs=mean(total_cs), sd.CpGs=sd(total_cs))

# Good samples
sample.info %>% filter(sample_seq %in%  c("1", "2", "3", "4", "7", "10", "11", "12", "15", "17", "18", "20")) %>%
  summarise(mean.aligned=mean(perc_totalread_unique), sd.aligned=sd(perc_totalread_unique),
            min.aligned=min(perc_totalread_unique), max.aligned=max(perc_totalread_unique),
            mean.reads=mean(total_reads), sd.reads=sd(total_reads),
            mean.CpGs=mean(total_cs), sd.CpGs=sd(total_cs))

# sample.info %>% arrange(sample_seq) %>% dplyr::select(sample_seq, sample) %>% 
#   filter(sample_seq %!in%  c("1", "2", "3", "4", "7", "10", "11", "12", "15", "17", "18", "20")) %>% write_clip()
```
### Is bacterial load high in "bad" samples?

```{r}
# specify the path to your files
path <- "../results/MEGAN"

# get all files in the directory
files <- list.files(path, pattern = "*.tab", full.names = TRUE) 

megan.df <- files %>%
  map_df(~{
    read_tsv(.x, col_names = FALSE) %>%
    set_names(c("phylum", "count")) %>%
    mutate(sample = str_remove(basename(.x), ".MEGAN.taxonName_to_count_summarized-phylum.tab"))  # remove the extension from the filename
  })

megan.df <- megan.df %>%
 mutate(sample.read = str_replace_all(sample, '.trimmed', "")) %>%
 mutate(sample = str_replace_all(sample, '.trimmed.R1|.trimmed.R2', "")) %>%
 left_join(
    read_sheet("https://docs.google.com/spreadsheets/d/1ym0XnYVts98tIUCn0kIaU6VuvqxzV7LoSx9RHwLdiIs/edit?usp=sharing", "data_071119collection") %>% 
 clean_names() %>% dplyr::select(sample, tank, high_or_low_co2, developmental_stage, dna_siolation_batch), 
            by="sample") %>%
  mutate_at(vars(phylum, sample, sample.read, tank, high_or_low_co2, developmental_stage, dna_siolation_batch), as.factor)

# # To determine top phyla
# options(scipen = 999)
# megan.df %>% 
#  pivot_wider(names_from = phylum, values_from = count, values_fill = 0) %>%
#  mutate(total = rowSums(dplyr::select(., -sample, -sample.read, -tank, -high_or_low_co2, -developmental_stage, -dna_siolation_batch), na.rm = TRUE)) %>%
#  mutate_at(vars(-sample, -sample.read, -tank, -high_or_low_co2, -developmental_stage, -dna_siolation_batch, -total), funs(. / total, )) %>%
#  # calculate mean fraction for each phylum to ID the top contributors
#  dplyr::select(where(is.numeric)) %>%
#   summarise(across(c(everything(), -total), mean)) %>% t() %>% as.data.frame() %>% arrange(desc(V1)) %>% round(digits=3)
# options(scipen = 0)

megan.df.plot <- megan.df %>%
 pivot_wider(names_from = phylum, values_from = count, values_fill = 0) %>%
 mutate(total = rowSums(dplyr::select(., -sample, -sample.read, -tank, -high_or_low_co2, -developmental_stage, -dna_siolation_batch), na.rm = TRUE)) %>%
 mutate_at(vars(-sample, -sample.read, -tank, -high_or_low_co2, -developmental_stage, -dna_siolation_batch, -total), funs(. / total, )) %>%
 arrange(desc(Arthropoda)) %>%
 mutate(sample.read = factor(sample.read, rev(unique(sample.read)), ordered = TRUE)) %>%
 pivot_longer(cols = c(-total, -sample, -sample.read, -tank, -high_or_low_co2, -developmental_stage, -dna_siolation_batch),
              names_to = "phylum", values_to = "fraction") %>%
  filter(phylum %in% c("Arthropoda", "Chordata", "Microsporidia", "Proteobacteria", "Bacteroidetes"))

# generate vector of colors identifying treatment for x axis labels 
a <- ifelse((megan.df.plot %>% dplyr::select(sample.read, high_or_low_co2) %>% distinct())$high_or_low_co2 == "H", "red", "blue")

ggplot(megan.df.plot, aes(fill=phylum, y=fraction, x=sample.read)) + geom_bar(position="stack", stat="identity") + 
  theme_minimal() + theme(legend.position = "bottom", axis.text.x = element_text(angle = 90, colour=rev(a))) +
  xlab(NULL) + ylab("Fraction of reads")

save(megan.df, file = "../results/MEGAN/megan.df")
save(megan.df.plot, file = "../results/MEGAN/megan.df.plot")
```

### Run analyses without the low-CpG methylation samples 
Good samples: 
control pCO2: 1, 2, 3, 4, 7, 12
high pCO2:    10, 11, 15, 17, 18, 20

#### Generate PCA without weird samples (those with very low meth rates)

```{r}
meth_unite_nolows <- reorganize(methylObj = meth_unite, 
                                sample.ids = c("1", "2", "3", "4", "7", "10", "11", "12", "15", "17", "18", "20"),
                                treatment = as.numeric((sample.info %>% 
                                                          arrange(sample_seq))$high_or_low_co2)[c(1:4, 7, 10, 11, 12, 15, 17, 18, 20)])
PCASamples(meth_unite_nolows)
```

```{r}
(sample.info %>% arrange(sample_seq))[c(1:4, 7, 10, 11, 12, 15, 17, 18, 20),]
```

```{r}
meth_obj_nolow <- reorganize(methylObj = meth_obj, sample.ids = c("1", "2", "3", "4", "7", "10", "11", "12", "15", "17", "18", "20"),
                             treatment = as.numeric((sample.info %>% 
                                                          arrange(sample_seq))$high_or_low_co2)[c(1:4, 7, 10, 11, 12, 15, 17, 18, 20)])

meth_5x=filterByCoverage(meth_obj_nolow,
                         lo.count=5,lo.perc=NULL,
                                      hi.count=100,hi.perc=NULL)

meth_unite_5x=methylKit::unite(meth_5x, destrand=TRUE, min.per.group=5L)
# For some reason the sample id's aren't accurately assigned to each column. 


PCA.5x <- PCASamples(meth_unite_5x, obj.return = TRUE) #Run a PCA analysis on percent methylation for all samples. methylKit uses prcomp to create the PCA matrix
summary(PCA.5x)  
#PC1=11.6% PC2=10.6% 

meth_10x=filterByCoverage(meth_obj_nolow,
                         lo.count=10,lo.perc=NULL,
                                      hi.count=100,hi.perc=NULL)

meth_unite_10x=methylKit::unite(meth_10x, destrand=TRUE)  # by not setting min.per.group manually, all samples must meet threshold for locus to be retained

save(meth_10x, file="../results/methylkit/meth_10x")
save(meth_unite_10x, file="../results/methylkit/meth_unite_10x")

load(file="../results/methylkit/meth_10x")
load(file="../results/methylkit/meth_unite_10x")
```
Important note!  The column names in the meth_unite_10x object are NOT the sample IDs. 

THE UNITE() FUNCTION ASSIGNS COLUMN NAMES USING THE CONSECUTIVE NUMBERS 1-13, WHILE THE SAMPLE ID'S ARE 1,2,3,4,7,10,11,12,15,17,18,20.  

```{r}
getSampleID(meth_unite_10x) #correct sample IDs
colnames(meth_unite_10x) # columns containing data do NOT have sample iDs, but are rather the n'th item in the vector of sample IDs 
```
PCA analysis & figure 

```{r}
PCA.10x <- PCASamples(meth_unite_10x, obj.return = TRUE) #Run a PCA analysis on percent methylation for all samples. methylKit uses prcomp to create the PCA matrix
summary(PCA.10x)  
#PC1=12.0% PC2=10.4% 
```


```{r}
# PCA using prcomp etc., but NOTE that I pull the covariance matrix from the SNPRelate PCA command  
# pca.princomp.gen <- princomp(covmat=pca$genmat, scores=T, cor = F) #specify "covmat" because input is a covariance matrix (not raw data!)
# summary(pca.princomp.gen)
# pca.eigenval(pca.princomp.gen)[2,1:5] #The Proporation of Variance (aka %variance)  
# screeplot(pca.princomp.gen, bstick=TRUE) #PC axes aren't super sign.


screeplot(PCA.10x, bstick=TRUE) #PC axes aren't super sign.

getSampleID(meth_unite_10x) #correct sample IDs

tab.meth <- data.frame(sample.id = rownames(PCA.10x$x),
    PC1.meth = PCA.10x$x[,1],    # the first eigenvector
    PC2.meth = PCA.10x$x[,2],    # the second eigenvector
    PC3.meth = PCA.10x$x[,3],    # the third eigenvector
    PC4.meth = PCA.10x$x[,4],    # the fourth eigenvector
    stringsAsFactors = FALSE)
pc.percent.meth <- summary(PCA.10x)$importance["Proportion of Variance",c("PC1", "PC2", "PC3", "PC4")]

shapiro.test(PCA.10x$x) #Multivariate normality assumption not met

tab.meth.annot <- left_join(tab.meth, sample.info %>% mutate(sample_seq=as.character(sample_seq)), 
                            by=c("sample.id"="sample_seq")) %>% droplevels()
group <- as.factor(tab.meth.annot$high_or_low_co2)

ggscatter(tab.meth.annot,
          x="PC1.meth", y="PC2.meth", col="high_or_low_co2", shape="high_or_low_co2", size=3.5, alpha=0.85, 
          ellipse = F, star.plot = T) +
  theme_minimal() + ggtitle("Global methylation PC1xPC2") + 
  xlab(paste("PC1 (", 100*round(pc.percent.meth[1], digits = 3), "%)", sep="")) + 
  ylab(paste("PC2 (", 100*round(pc.percent.meth[2], digits = 3), "%)", sep="")) + 
  theme(legend.position = "right", legend.text=element_text(size=8), legend.title=element_text(size=9)) + 
  scale_color_manual(name="Treatment", 
                     values=c("L"="dodgerblue3",
                              "H"="firebrick3"),
                     labels=c("L"="Control",
                              "H"="High pCO2")) +
  scale_shape_manual(guide="none", values=c("L"=19, "H"=17)) +
  guides(fill="none", colour = guide_legend(override.aes = list(size=3.5, linetype="blank", shape=c(19, 17))))
```
<!-- PCA plot using ordiplot -->

<!-- ```{r} -->
<!-- plotCustomization <- (sample.info %>% arrange(sample_seq))[c(1:4, 7, 10, 11, 12, 15, 17, 18, 20),c("high_or_low_co2", "sample_seq", "tank")] %>% -->
<!--   mutate(color=case_when(high_or_low_co2 == "H" ~ "firebrick3", TRUE ~ "dodgerblue3"), -->
<!--          symbol=case_when(high_or_low_co2 == "H" ~ 16, TRUE ~ 17)) -->

<!-- # Save size: width=650 height=600 -->
<!-- par(mar = c(5, 5, 1, 1)) #Specify inner and outer margins -->
<!-- PCA.figure <- ordiplot(PCA.10x, choices = c(1, 2), type = "none", display = "sites", cex = 0.4, xlab = "", ylab = "", xaxt = "n", yaxt = "n") #Use ordiplot to create base biplot. Do not add any points -->
<!-- points(PCA.figure, "sites", col = as.character(plotCustomization$color), pch = plotCustomization$symbol, cex = 2.2)  #use points -->
<!-- #text(PCA.figure$sites, col = as.character(plotCustomization$color), labels=plotCustomization$sample_seq, cex=1.5) #use sample number  -->

<!-- #Add multiple white boxes on top of the default black box to manually change the color -->
<!-- box(col = "white") -->
<!-- box(col = "white") -->
<!-- box(col = "white") -->
<!-- box(col = "white") -->
<!-- box(col = "white") -->
<!-- box(col = "white") -->
<!-- box(col = "white") -->
<!-- box(col = "white") -->
<!-- box(col = "white") -->
<!-- box(col = "white") -->
<!-- #ordiellipse(PCA.filtered, plotCustomization$population, show.groups = "HC", col = "firebrick3") #Add confidence ellipse around the samples in elevated pCO2 -->
<!-- #ordiellipse(PCA.filtered, plotCustomization$population, show.groups = "SS", col = "dodgerblue3") #Add confidence ellipse around the samples in ambient pCO2 -->
<!-- axis(side =  1, labels = TRUE, col = "grey80", cex.axis = 1.7) #Add x-axis -->
<!-- mtext(side = 1, text = "PC 1 (12.0%)", line = 3, cex = 1.5) #Add x-axis label -->
<!-- axis(side =  2, labels = TRUE, col = "grey80", cex.axis = 1.7) #Add y-axis -->
<!-- mtext(side = 2, text = "PC 2 (10.4%)", line = 3, cex = 1.5) #Add y-axis label -->
<!-- legend("topleft",  -->
<!--        pch = c(16, 17),  -->
<!--        legend = c("High pCO2", "Control pCO2"),  -->
<!--        col = c(c("firebrick3", "dodgerblue3")),  -->
<!--        cex = 1.3, bty = "n") #Add a legend with information about ambient and elevated samples -->
<!-- ``` -->

### Save PC scores and R object for possible integration with other data

NOTE: this PCA was run with _all_ methylated loci data after filtering for 10x across all retained samples. 

```{r}
save(PCA.10x, file="../results/methylkit/PCA.10x")
write_delim(data.frame(PCA.10x$x) %>% tibble::rownames_to_column("sample"), "../results/methylkit/PC-scores-methylation-10x.tab",  delim = '\t', col_names = TRUE)
```


### Quick check to confirm that removed samples should definitely be removed: 

Filter for 10x across ALL samples (this is including the low-CpG methylation samples) to assess clustering. 
Yes, the weird samples still are separated along PC1, which explains the vast majority of variation (68%), from the other samples. Removal of those samples is confirmed. 

```{r}
meth_10x_allsamples=filterByCoverage(meth_obj,
                         lo.count=10,lo.perc=NULL,
                                      hi.count=100,hi.perc=NULL)

meth_unite_10x_allsamples=methylKit::unite(meth_10x_allsamples, destrand=TRUE, min.per.group=10L) #require all 10 samples to have 10x
PCA.10x_allsamples <- PCASamples(meth_unite_10x_allsamples, obj.return = TRUE) #Run a PCA analysis on percent methylation for all samples. methylKit uses prcomp to create the PCA matrix
summary(PCA.10x_allsamples)  
#PC1=68.0% PC2=6.8% 
```

```{r}
plotCustomization_allsamples <- (sample.info %>% arrange(sample_seq))[,c("high_or_low_co2", "sample_seq", "tank")] %>%
  mutate(color=case_when(high_or_low_co2 == "H" ~ "firebrick3", TRUE ~ "dodgerblue3"),
         symbol=case_when(high_or_low_co2 == "H" ~ 16, TRUE ~ 17))

# Save size: width=650 height=600
par(mar = c(5, 5, 1, 1)) #Specify inner and outer margins
PCA.figure_allsamples <- ordiplot(PCA.10x_allsamples, choices = c(1, 2), type = "none", display = "sites", cex = 0.4, xlab = "", ylab = "", xaxt = "n", yaxt = "n") #Use ordiplot to create base biplot. Do not add any points
#points(PCA.figure_allsamples, "sites", col = as.character(plotCustomization_allsamples$color), pch = plotCustomization_allsamples$symbol, cex = 2.5)  #use points
text(PCA.figure_allsamples$sites, col = as.character(plotCustomization_allsamples$color), labels=plotCustomization_allsamples$sample_seq, cex=1.3) #use tank number 

#Add multiple white boxes on top of the default black box to manually change the color
box(col = "white")
box(col = "white")
box(col = "white")
box(col = "white")
box(col = "white")
box(col = "white")
box(col = "white")
box(col = "white")
box(col = "white")
box(col = "white")
#ordiellipse(PCA.filtered, plotCustomization$population, show.groups = "HC", col = "firebrick3") #Add confidence ellipse around the samples in elevated pCO2
#ordiellipse(PCA.filtered, plotCustomization$population, show.groups = "SS", col = "dodgerblue3") #Add confidence ellipse around the samples in ambient pCO2
axis(side =  1, labels = TRUE, col = "grey80", cex.axis = 1.7) #Add x-axis
mtext(side = 1, text = "PC 1 (60%)", line = 3, cex = 1.5) #Add x-axis label
axis(side =  2, labels = TRUE, col = "grey80", cex.axis = 1.7) #Add y-axis
mtext(side = 2, text = "PC 2 (9.5%)", line = 3, cex = 1.5) #Add y-axis label
legend("bottomright", 
       pch = c(16, 17), 
       legend = c("High pCO2", "Low pCO2"), 
       col = c(c("firebrick3", "dodgerblue3")), 
       cex = 1.3, bty = "n") #Add a legend with information about ambient and elevated samples
```

```{r, warning=FALSE}
getCorrelation(meth_unite_10x_allsamples,plot=TRUE)
```

### Hierarchical Clustering using filtered methylation data

The function clusters samples using hclust function and various distance metrics derived from percent methylation per base or per region for each sample.

```{r}
clusterSamples(meth_unite_10x, dist="correlation", method="ward", plot=TRUE)
```


```{r}
getData(meth_unite_5x) %>% nrow() #340,394 loci @ 5x coverage
getData(meth_unite_10x) %>% nrow() #146,761 loci @ 10x coverage
```

Percent methylation matrix (rows=region/base, columns=sample) can be extracted from methylBase object by using percMethylation function. This can be useful for downstream analyses. 

Here I create % methylation matrices from the filtered object, and use it to do another cluster analysis 

```{r}
perc.meth=percMethylation(meth_unite_10x, rowids=T)
hist(perc.meth, col="gray50" )

# # Save % methylation df to object and .tab file 
# save(perc.meth, file = "../analyses/methylation-filtered/R-objects/perc.meth") #save object to file 
#load(file = "../analyses/methylation-filtered/R-objects/perc.meth") #load object if needed
# write.table((as.data.frame(perc.meth) %>% tibble::rownames_to_column("contig")), file = here::here("analyses", "methylation-filtered", "percent-methylation-filtered.tab"), sep = '\t', na = "NA", row.names = FALSE, col.names = TRUE)

# What percentage of loci have ZERO methylation? 
100*table(perc.meth==0)[2]/table(perc.meth==0)[1] 
# 1.8% of loci unmethylated, averaged across all samples 

# Mean % methylation across all samples 
mean(perc.meth, na.rm=TRUE) 
#79.2%

sample.info %>% dplyr::select(high_or_low_co2, sample_seq) %>% arrange(sample_seq)
(perc.meth %>% as.matrix())[,c("1","2","3","4","7","12")] %>% head() # low pCO2 
(perc.meth %>% as.matrix())[,c("10","11","15", "17", "18", "20")] %>% head() # high pCO2 

# Mean % methylation for each treatment 
mean((perc.meth %>% as.matrix())[,c("1","2","3","4","7","12")], na.rm=TRUE) # Low CO2 exposed - 79.5%
mean((perc.meth %>% as.matrix())[,c("10","11","15", "17", "18", "20")], na.rm=TRUE) # High CO2 exposed - 78.9% 

perc.meth.summ <- data.frame(sample=1:12, methylated=1:12, coverage=1:12, mean.perc=1:12)
for (i in 1:12) {
  perc.meth.summ[i,"sample"] <- getSampleID(meth_unite_10x)[i]
  perc.meth.summ[i,"methylated"] <-mean(as.vector(getData(meth_unite_10x)[,grep(c("numCs"), colnames(meth_unite_10x))][i][,1]), na.rm=T)
  perc.meth.summ[i,"coverage"] <- mean(as.vector(getData(meth_unite_10x)[,grep(c("coverage"), colnames(meth_unite_10x))][i][,1]), na.rm=T)
  perc.meth.summ[i,"mean.perc"] <- mean(as.vector(getData(meth_unite_10x)[,grep(c("numCs"), colnames(meth_unite_10x))][i][,1]) / 
                                     as.vector(getData(meth_unite_10x)[,grep(c("coverage"), colnames(meth_unite_10x))][i][,1]), na.rm=T)
}
mean(perc.meth.summ[perc.meth.summ$sample %in% c("1","2","3","4","7","12"),"mean.perc"]) #Loe pCO2 treatment = 79.5%
mean(perc.meth.summ[perc.meth.summ$sample %in% c("10","11","15", "17", "18", "20"),"mean.perc"]) #78.9%

paste("No. of loci in filtered dataset:", nrow(perc.meth)) #146,761 loci 
paste("No. of samples:", ncol(perc.meth)) #12 sampes 

# % CpG Methylation by sample after filtering for 10x coverage (i.e. uniting)
perc.meth.summ[,c("sample", "mean.perc")] %>% mutate(mean.perc=round(100*mean.perc, digits = 1)) %>% View()
```

```{r}
myDiff=calculateDiffMeth(meth_unite_10x)
```

```{r}
# get  deferentially methylated bases using 12% difference and save to files - 12,971 loci 
myDiff12p=getMethylDiff(myDiff,difference=12,qvalue=0.01)
paste("Number diff. methylated loci with 12% difference: ", getData(myDiff12p) %>% nrow())

# get all differentially methylated bases - 1,427 loci
myDiff25p=getMethylDiff(myDiff,difference=25,qvalue=0.01)
paste("Number diff. methylated loci with 25% difference: ", getData(myDiff25p) %>% nrow())

# Extract hypo and hyper methylated loci. Here "type" must think that the low pCO2 is the experimental treatment, since hypo and hyper are switched. 
# get hypo methylated bases - 843 loci
myDiff25p.hyper=getMethylDiff(myDiff,difference=25,qvalue=0.01,type="hyper") 
paste("Number loci hypo-methylated in high pCO2 with 25% difference: ", getData(myDiff25p.hyper) %>% nrow())

# get hypo methylated bases - 593 loci 
myDiff25p.hypo=getMethylDiff(myDiff,difference=25,qvalue=0.01,type="hypo")
paste("Number loci hyper-methylated in high pCO2 with 25% difference: ", getData(myDiff25p.hypo) %>% nrow())
  
#write.table(myDiff12p, file = "../analyses/DMLs/myDiff12p.tab", sep = "\t", col.names = TRUE)
#write_delim(getData(myDiff12p) %>%  mutate(start = start-1, end = end+1) %>% dplyr::select(chr, start, end, meth.diff),
#            "../analyses/DMLs/dml12.bed",  delim = '\t', col_names = TRUE) 
```

```{r}
write.table(myDiff25p, file = "../results/methylkit/myDiff25p.tab", sep = "\t", col.names = TRUE)
save(myDiff25p, file="../results/methylkit/myDiff25p")
#load(file="../results/methylkit/myDiff25p") #load if needed
```


PCA analysis & figure, DML's only for integration with expression data! 

```{r}
temp <- meth_unite_10x %>% set_rownames(paste(meth_unite_10x$chr, meth_unite_10x$start, sep="."))
dmls <- c((myDiff25p %>% getData() %>% mutate(dml=paste(chr, start, sep=".")))$dml)

PCA.10x.dmls <- PCASamples(temp[rownames(temp) %in% dmls,], obj.return = TRUE)
summary(PCA.10x.dmls)  
#PC1=26.2% PC2=9.3% 

screeplot(PCA.10x.dmls, bstick=TRUE) #PC1 significant 

tab.meth.dml <- data.frame(sample.id = rownames(PCA.10x.dmls$x),
    PC1.meth = PCA.10x.dmls$x[,1],    # the first eigenvector
    PC2.meth = PCA.10x.dmls$x[,2],    # the second eigenvector
    PC3.meth = PCA.10x.dmls$x[,3],    # the third eigenvector
    PC4.meth = PCA.10x.dmls$x[,4],    # the fourth eigenvector
    stringsAsFactors = FALSE)
pc.percent.meth.dmls <- summary(PCA.10x.dmls)$importance["Proportion of Variance",c("PC1", "PC2", "PC3", "PC4")]

shapiro.test(PCA.10x.dmls$x) #Multivariate normality assumption met

tab.meth.dmls.annot <- left_join(tab.meth.dml, sample.info %>% mutate(sample_seq=as.character(sample_seq)), 
                            by=c("sample.id"="sample_seq")) %>% droplevels()
group <- as.factor(tab.meth.dmls.annot$high_or_low_co2)

ggscatter(tab.meth.dmls.annot,
          x="PC1.meth", y="PC2.meth", col="high_or_low_co2", shape="high_or_low_co2", size=3.5, alpha=0.85, 
          ellipse = F, star.plot = T) +
  theme_minimal() + ggtitle("Global methylation PC1xPC2") + 
  xlab(paste("PC1 (", 100*round(pc.percent.meth.dmls[1], digits = 3), "%)", sep="")) + 
  ylab(paste("PC2 (", 100*round(pc.percent.meth.dmls[2], digits = 3), "%)", sep="")) + 
  theme(legend.position = "right", legend.text=element_text(size=8), legend.title=element_text(size=9)) + 
  scale_color_manual(name="Treatment", 
                     values=c("L"="dodgerblue3",
                              "H"="firebrick3"),
                     labels=c("L"="Control",
                              "H"="High pCO2")) +
  scale_shape_manual(guide="none", values=c("L"=19, "H"=17)) +
  guides(fill="none", colour = guide_legend(override.aes = list(size=3.5, linetype="blank", shape=c(19, 17))))
```



### Take the DMLs to a bed

```{r}
# Write out bedfile for DMLs 
dml25 <- myDiff25p %>% getData() %>% 
  mutate(start = start-1, end = end+1) %>% 
  dplyr::select(chr, start, end, meth.diff)

write_delim(dml25, "../results/methylkit/dml25.bed",  delim = '\t', col_names = F)
write_delim(dml25, "../results/methylkit/dml25_forIGV.bed",  delim = '\t', col_names = FALSE) #no column names for IGV
save(dml25, file="../results/methylkit/dml25")
#load(file="../results/methylkit/dml25") #load dml25 if needed

# Write out bedfile for all loci fed into DML analysis, to use as background for enrichment 
mydiff.all <-  mutate(getData(myDiff), start = start -1, end = end + 1) %>% dplyr::select(chr, start, end, meth.diff) %>% mutate_if(is.numeric, as.integer)
write_delim(mydiff.all, "../results/methylkit/mydiff-all.bed",  delim = '\t', col_names = FALSE)
save(mydiff.all, file="../results/methylkit/mydiff.all")
#load(file="../results/methylkit/mydiff.all")
```

```{r}
paste("No. of DMLs that had higher methylation in high pCO2 = ", dml25 %>%filter(meth.diff<0)%>%nrow(), " (", 100*round(dml25 %>%filter(meth.diff<0)%>%nrow()/nrow(dml25), digits = 2), "%)", sep="") 
paste("No. of DMLs that had lower methylation in high pCO2= ", dml25 %>%filter(meth.diff>0)%>%nrow(), " (", 100*round(dml25 %>%filter(meth.diff>0)%>%nrow()/nrow(dml25), digits = 2), "%)", sep="") 

# Save DML bed files with hyper- and hypo-methylated loci in high pCO2 treatment 
# meth.diff < 0 = loci where methylation rate is HIGHER in high pCO2 treatment 

write_delim(dml25 %>%filter(meth.diff<0), "../results/methylkit/dml25-hypermeth_in_OA.bed",  delim = '\t', col_names = TRUE) 
write_delim(dml25 %>%filter(meth.diff>0), "../results/methylkit/dml25-hypometh_in_OA.bed",  delim = '\t', col_names = TRUE) 
```

Plot mean % methylation for loci by those that hypo- and hyper-methylated in OA-exposed crab 

```{r}
meth_unite_reshaped %>%
  filter(coverage>=10) %>%
  filter(sample %in% c(1,2,3,4,7,12, #low pCO3
                       8,10,11,15,17,18,20)) %>% #high pCO2
  mutate(locus=paste(chr, start-1, sep=".")) %>%  # Create locus column; subtract 1 from the "start" column to match dml25
  filter(locus %in% (dml25 %>%filter(meth.diff<0) %>% mutate(locus=paste(chr, start, sep=".")))$locus) %>% #keep loci with meth.diff < 0, higher meth in high pCO2
  left_join(sample.info[,c("high_or_low_co2", "sample_seq")], by=c("sample"="sample_seq")) %>%
  group_by(high_or_low_co2, locus) %>%
  summarise(mean = mean(percMeth)) %>%
  ggplot(aes(x=high_or_low_co2, y=mean)) + geom_violin() + geom_jitter() + 
  theme_minimal() + xlab("pCO2 treatment") +
  ggtitle("Loci with higher methylation in OA (n=593)")


#dml25 %>%filter(meth.diff>0) %>% nrow()

meth_unite_reshaped %>%
  filter(coverage>=10) %>%
  filter(sample %in% c(1,2,3,4,7,12, #low pCO3
                     8,10,11,15,17,18,20)) %>% #high pCO2
  mutate(locus=paste(chr, start-1, sep=".")) %>%  # Create locus column; subtract 1 from the "start" column to match dml25
  filter(locus %in% (dml25 %>%filter(meth.diff>0) %>% mutate(locus=paste(chr, start, sep=".")))$locus) %>% #keep loci with meth.diff > 0, lower meth in high pCO2
  left_join(sample.info[,c("high_or_low_co2", "sample_seq")], by=c("sample"="sample_seq")) %>%
  group_by(high_or_low_co2, locus) %>%
  summarise(mean = mean(percMeth)) %>%
  ggplot(aes(x=high_or_low_co2, y=mean)) + geom_violin() + geom_jitter() + 
  theme_minimal() + xlab("pCO2 treatment") +
  ggtitle("Loci with lower methylation in OA (n=834)")
```
###  Subset the filtered methylation count dataframe for only those that are differentially methylated. Save object to file. 

```{r}
dml25_counts <- selectByOverlap(meth_unite_10x,as(myDiff25p,"GRanges"))
class(dml25_counts) #class should be methylBase
perc.methDMLs= percMethylation(dml25_counts, rowids=T) #create a percent methylated object for DMLs

# save count and percent objects to be used in subsequent notebook 
save(dml25_counts, file = "../results/methylkit/dml25_counts")
save(perc.methDMLs, file="../results/methylkit/perc.methDMLs")

#load(file = "../results/methylkit/dml25_counts")
#load(file="../results/methylkit/perc.methDMLs")
```
## permANOVA and dispersion test, % methylation by treatment 

```{r}
perc.meth.d<-vegdist(data.frame(t(perc.meth)),"euclidean", na.rm = TRUE)
meth.perm <- adonis(data.frame(t(perc.meth)) ~ high_or_low_co2, data=plotCustomization,
                          permutations=1000, method='euclidean', na.rm = TRUE) #yes, global population difference

meth.perm$aov.tab # 

hist(meth.perm$f.perms, main="Histogram of pseudo-F values under the null model",
     xlab="pseudo=F values", col="gray", xlim=c(0.5,2.5))
abline(v=1.9743, col="red") # indicates F-value is significant.

source("../references/biostats.R")
pca.eigenval(PCA.10x) #The Proporation of Variance = %variance explained
screeplot(PCA.10x, bstick=TRUE) #that's weird- PC's 1-5 aren't sign. but the higher PCs are. 

# # check sign. of PCs - can't get this to work, though.
# ordi.monte(PCA.10x, ord="pca", dim=2, perm = 1000, scale = F, plot = F)

anova(meth.bdp<-betadisper(perc.meth.d,plotCustomization$high_or_low_co2)) #no sign. difference in dispersion, which is good
boxplot(meth.bdp)

# # perform PCoA just for fun on distance matrix 
# meth.pcoa<-cmdscale(perc.meth.d, eig=TRUE, add=T)
# ordiplot(meth.pcoa, choices = c(1, 2), type="text", display="sites", xlab="PC 1", ylab="PC 2")
```

### Reshape filtered data to create a new column with sample info, and calculate %methylation for each locus/sample 

```{r}
# Get sample ID vector from methylBase object (created from unite() function)
samples <- getSampleID(meth_unite_10x)

meth_unite_10x_reshaped <- melt(data=meth_unite_10x, id=c("chr", "start", "end", "strand"), value.name = "count") %>%
 tidyr::separate(col = "variable" , into = c("variable", "sample_temp"), sep = "(?<=[a-zA-Z])\\s*(?=[0-9])") %>%
  dcast(chr+start+end+strand+sample_temp ~  variable) %>%

  #Correct sample IDs extracted from unite() function 
  mutate(sample=case_when(
    sample_temp== 1 ~ samples[1],
    sample_temp== 2 ~ samples[2],
    sample_temp== 3 ~ samples[3],
    sample_temp== 4 ~ samples[4],
    sample_temp== 5 ~ samples[5],
    sample_temp== 6 ~ samples[6],
    sample_temp== 7 ~ samples[7],
    sample_temp== 8 ~ samples[8],
    sample_temp== 9 ~ samples[9],
    sample_temp== 10 ~ samples[10],
    sample_temp== 11 ~ samples[11],
    sample_temp== 12 ~ samples[12],
    sample_temp== 13 ~ samples[13])) %>%
  dplyr::select(-sample_temp) %>%
  
  mutate(percMeth = 100*(numCs/coverage)) %>% 
  left_join(sample.info[c("sample_seq", "high_or_low_co2")] %>% 
              mutate(sample_seq=as.character(sample_seq)), by=c("sample"="sample_seq")) %>%
  dplyr::rename("treatment"="high_or_low_co2")

save(meth_unite_10x_reshaped, file="../results/methylkit/meth_unite_10x_reshaped")
#load(file="../results/methylkit/meth_unite_10x_reshaped")

# save to tab file if needed
#readr::write_delim(meth_unite_10x_reshaped, "../results/methylkit/meth_unite_10x_reshaped.tab"),  delim = '\t', col_names = FALSE)
```

## Check out overall distribution of filtered methylation data by  treatment 

Any trends? No, similar distribution of % CpG methyation 

```{r}
meth_unite_10x_reshaped %>% 
  ggplot(aes(x= treatment, y=percMeth, fill= treatment)) +
  geom_violin(width=1, alpha=0.8) + theme_minimal() +
  ylab("% CpG Methylation") + theme(axis.title.x = element_blank(), legend.position = "none",
                                    text = element_text(size=16)) +
  scale_x_discrete(labels=c("H" = "High pCO2", "L" = "Control pCO2")) + 
  scale_fill_manual(values=c("firebrick3", "dodgerblue3"))
```

View distributions in density plot 

Looks as though there is slightly more loci with higher methylation rates in the High pCO2 treatment (red distribution is shifted to the right a bit)

```{r}
#ggplotly(
meth_unite_10x_reshaped %>% 
  ggplot(aes(x=percMeth, fill= treatment)) +
  geom_density(alpha=0.5) + 
  scale_fill_manual(values=c("firebrick3", "dodgerblue3"), 
                    labels=c("H" = "High pCO2", "L" = "Control pCO2"), name=NULL) +
  #xlim(90,100) + #use to find location on x axis of second highest peak
  theme_minimal() +
  ylab("Density") + xlab("% CpG Methylation") +
  theme(text = element_text(size=18), legend.position = "top") #)
```

Find location of peaks in each treatment in the above figure. Used this as a reference: http://ianmadd.github.io/pages/PeakDensityDistribution.html

Answer:  
- Largest peak for high pCO2 treatment is at 99.98% 
- Largest peak for low pCO2 treatment (control) is 94.56% 

```{r}
## Find largest peak 

# Find largest peak for High pCO2
density(subset(meth_unite_10x_reshaped,  treatment=="H" & percMeth!="NA")$percMeth) # Y maximum is 4.780e-02

# Find where that maximum lies (this gives the index #)
which.max(density(subset(meth_unite_10x_reshaped,  treatment=="H" & percMeth!="NA")$percMeth)$y) #answer = index #497

# Find the x value where the largest Y value is
density(subset(meth_unite_10x_reshaped,  treatment=="H" & percMeth!="NA")$percMeth)$x[497] #H max (4.895e-02) is @ 99.90%

##--------------

# Find largest peak for Low pCO2, i.e. control
density(subset(meth_unite_10x_reshaped,  treatment=="L" & percMeth!="NA")$percMeth) # Y maximum is 4.663e-02  

which.max(density(subset(meth_unite_10x_reshaped,  treatment=="L" & percMeth!="NA")$percMeth)$y) #answer = 472

density(subset(meth_unite_10x_reshaped,  treatment=="L" & percMeth!="NA")$percMeth)$x[472] #L max (0.0469560  ) is 94.52%

## --------------

## Find 2nd largest peak. NOTE: I looked at the ggplotly density figure above and noted that the 2nd peaks end before ~98%, so I subset the percMeth data to retain only those <98%, then find the peak 

# Find 2nd largest peak for High pCO2
density(subset(meth_unite_10x_reshaped,  treatment=="H" & percMeth!="NA" & percMeth<98)$percMeth) # Y maximum is .04997

# Find where that maximum lies (this gives the index #)
which.max(density(subset(meth_unite_10x_reshaped,  treatment=="H" & percMeth!="NA" & percMeth<98)$percMeth)$y) #answer = index #478

# Find the x value where the largest Y value is
density(subset(meth_unite_10x_reshaped,  treatment=="H" & percMeth!="NA" & percMeth<98)$percMeth)$x[478] #H max (5.281e-02) is @ 94.40%

#---
# Find 2nd largest peak for Low pCO2
density(subset(meth_unite_10x_reshaped,  treatment=="L" & percMeth!="NA" & percMeth>95)$percMeth) # Y maximum is 1.1686

# Find where that maximum lies (this gives the index #)
which.max(density(subset(meth_unite_10x_reshaped,  treatment=="L" & percMeth!="NA" & percMeth>95)$percMeth)$y) #answer = index #475

# Find the x value where the largest Y value is
density(subset(meth_unite_10x_reshaped,  treatment=="L" & percMeth!="NA" & percMeth>95)$percMeth)$x[475] #L max (4.953e-02) is @ 100.00%

# Very similar 2nd peak CpG methylation - High pCO2 = 94.24%, and Low pCO2 = 94.61% (only slightly higher in low pCO2 treatment)

#----- Double check # of loci 
meth_unite_10x_reshaped %>%
  filter( treatment=="H" & percMeth!="NA") %>%
  dplyr::select(chr, start) %>% unique() %>% nrow() #146,761

meth_unite_10x_reshaped %>%
  filter( treatment=="L" & percMeth!="NA") %>%
  dplyr::select(chr, start) %>% unique() %>% nrow() #146,761 (good they are the same)

#----- average % meth by pop

# High pCO2 mean = 78.92%, median = 88.27%
meth_unite_10x_reshaped %>%
  filter( treatment=="H" & percMeth!="NA") %>%
  dplyr::select(percMeth) %>% summary()

# Low (control) pCO2 mean = 79.54%, median = 88.71%
meth_unite_10x_reshaped %>%
  filter( treatment=="L" & percMeth!="NA") %>%
  dplyr::select(percMeth) %>% summary()
```

Takeaway = high and low (control) pCO2 have similar %CpG distributions on average, but the distribution of %meth varies a bit - both treatments' % methylation peak around 94% and 100%, but in the high pCO2 treatment the 100% peak is higher, and the 94% peak is lower. So, high pCO2 has more instances of 100% methylation. 

slightly higher methylation rates in the high pCO2 reared crab. 


### Calculations of % methylation by  treatment 

```{r}
# mean % methylation across samples  
dml25_counts_unique <- unique(getData(dml25_counts)[,c("chr", "start")]) %>%
  mutate(chr_start=paste(chr, start, sep="_"))

#Calculate ratio between mean % methylation in High pCO2 : Low pCO2 
DML.ratios <- meth_unite_10x_reshaped %>% 
  mutate(chr_start=paste(chr, start, sep="_")) %>%
  filter(chr_start %in% dml25_counts_unique$chr_start) %>% 
  group_by( treatment, chr, start) %>%
  summarise(median_percentMeth = median(percMeth, na.rm = TRUE)) %>% #mean_percentMeth = mean(percMeth, na.rm = TRUE), 
  dcast(chr + start ~  treatment) %>% 
  mutate(ratio_H.L = H/L,  #in this ratio column, values <1 = High pCO2 hypomethylated, values >1 = Low pCO2 hypomethylated 
         chr_start=paste(chr, start, sep="_"),
         diff_H.L = H-L)   #this is the difference between High pCO2 & Low pCO2 (H - L) 

nrow(DML.ratios)==nrow(dml25_counts_unique) #confirm same # loci; should=TRUE
save(DML.ratios, file="../results/methylkit/DML.ratios")
#load(file="../results/methylkit/DML.ratios")

# For all loci, calculate mean and sd percent methylation across samples by  treatment 
DML.calcs <- meth_unite_10x_reshaped %>% 
  group_by( treatment, chr, start) %>% 
  dplyr::summarise(
    mean_percMeth = mean(percMeth, na.rm=TRUE),
    sd_percMeth=sd(percMeth, na.rm=TRUE),
    n())  %>% 
  filter(chr %in% c(DML.ratios$chr) & 
           start %in% c(DML.ratios$start))
```

Interesting! DMLs on average have lower % methylation in the high pCO2 treatment 

```{r}
DML.ratios %>% 
pivot_longer(cols = c(H, L),
   names_to = "treatment",
   values_to = "mean_meth") %>%
  mutate(treatment=as.factor(treatment)) %>%
  ggplot(aes(x=treatment, y=mean_meth, fill=treatment)) +
  geom_violin(alpha=0.2) + geom_boxplot(alpha=0.4) +
  scale_fill_manual(values=c("firebrick3", "dodgerblue3"), 
                    labels=c("H" = "High pCO2", "L" = "Control pCO2"), name=NULL) + 
theme_minimal() + ggtitle("Methylation rates of DMLs") + 
  ylab("Methylation (mean within treatment)") + xlab(NULL)


DML.ratios %>% 
pivot_longer(cols = c(H, L),
   names_to = "treatment",
   values_to = "mean_meth") %>%
  mutate(treatment=as.factor(treatment)) %>%
  group_by(treatment) %>%
  summarise(mean=mean(mean_meth), 
             median=median(mean_meth),)
```

### For ALL loci, calculate difference between mean % methylation in H : L for divergent bar plot 

```{r}
#Calculate ratio between mean % methylation in High pCO2 : Low pCO2 (i.e. control) 
all.loci.ratios <- meth_unite_10x_reshaped %>% 
  mutate(chr_start=paste(chr, start, sep="_")) %>%
  group_by( treatment, chr, start) %>%
  summarise(mean_percentMeth = mean(percMeth, na.rm = TRUE)) %>% #could instead use ... allCs_percent = 100*(sum(numCs)/sum(coverage)), 
  dcast(chr + start ~  treatment) %>% 
  mutate(ratio_H.L = H/L,  #in this ratio column, values <1 = HC hypomethylated, values >1 = SS hypomethylated 
         chr_start=paste(chr, start, sep="_"),
         diff_H.L = H-L)   #this is the difference between H & L (H - L) 

# Any majuor differences in global methylatin among the  treatments? Here are summarys tats of % methylation across all shared loci: 
summary(all.loci.ratios$H) 
summary(all.loci.ratios$L)

# As determined in previous code chunks % meth is slightly higher in the high pCO2 treatment group, while % meth is slightly lower in DMLs in high pCO2 group 

# Do %methylation (all loci) distributions differ among  treatment using all loci? 
#ggplotly(
all.loci.ratios %>% 
pivot_longer(cols = c(H, L),
   names_to = "treatment",
   values_to = "mean_meth") %>%
  ggplot(aes(x=mean_meth, fill= treatment)) +
  geom_density(alpha=0.2) +
    scale_fill_manual(values=c("firebrick3", "dodgerblue3"), 
                    labels=c("H" = "High pCO2", "L" = "Control pCO2"), name=NULL) + 
theme_minimal() #)

# Using DMLs
DML.ratios %>% 
pivot_longer(cols = c(H, L),
   names_to = "treatment",
   values_to = "mean_meth") %>%
  ggplot(aes(x=mean_meth, fill= treatment)) +
  geom_density(alpha=0.2) +
    scale_fill_manual(values=c("firebrick3", "dodgerblue3"), 
                    labels=c("H" = "High pCO2", "L" = "Control pCO2"), name=NULL) + 
theme_minimal() + xlab("Mean methylation (within treatment)") + 
  ggtitle("% Methylation distribution for DMLs by treatment")
```

### Heatmap

```{r}
#load(file="../results/methylkit/meth_unite_10x_reshaped")
meth_unite_10x_reshaped %>% dplyr::select(sample, treatment) %>% distinct()

DMLs_heatmap <- 
  meth_unite_10x_reshaped %>% 
    left_join(sample.info[c("sample_seq", "tank")] %>%  mutate(sample_seq=as.character(sample_seq)), by=c("sample"="sample_seq")) %>% 
    arrange(treatment, tank) %>%
  mutate(sample=factor(sample, ordered = T, levels = c(3,4,7,1,2,12,20,17,8,10,18,11,15)), 
         chr_start=paste(chr, start, sep="_")) %>%
  filter(
           start %in% dml25_counts_unique$start & 
           chr %in% DML.ratios$chr)

#ggplotly(
ggplot(DMLs_heatmap, aes(sample, chr_start, fill= percMeth)) + xlab("Sample") + geom_tile(na.rm = T) +
  scale_y_discrete(limits=(DML.ratios[order(DML.ratios$ratio_H.L),]$chr_start)) + 
  #scale_fill_viridis(discrete=FALSE) 
  theme(axis.title.y=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank()) + 
  scale_fill_distiller(palette = "YlGnBu", direction = 1) #)
  #scale_fill_gradient(low="gray5", high="white")

sample.info[c("high_or_low_co2", "sample_seq", "tank")]
```

### Functional analysis of DMLs 

Identify genes that contain DMLs and run DAVID enrichment analysis

#### Load genesas results 

```{r}
require(janitor)
gensas <- read_delim(file = "../results/GenSAS/Metacarcinus-magister-v1.0.a1.6448195bde6ca-publish.genes.gff3", delim = "\t", skip = 5, 
                           col_names = c("seqname", "source", "feature", "start", "end", "score", "strand", "frame", "attribute")) %>%
  separate(attribute, sep = ";", remove=F, into = c("ID","Name","Parent")) %>%
  mutate_at(vars(start, end, score), as.numeric) %>%
  mutate_at(vars(feature), as.factor) %>%
  
#  group_by(SeqID, Start, End) %>% dplyr::slice(which.min(evalue))  %>% # where multiple blast hits for same gene, select one with minimum e-value
    mutate(ID=gsub("ID\\=", "", ID),
           Name=gsub("Name\\=", "", Name),
           Parent=gsub("Parent\\=", "", Parent)) %>%
  clean_names()

gensas.genes <- gensas %>% filter(feature=="gene")

gensas.blast <- read_delim(file = "../results/GenSAS/Metacarcinus-magister-v1.0.a1.640639c029cb6-blast_functional.tab", delim = "\t", skip = 12, col_names = T) %>%
  clean_names() %>%
  mutate(name=gsub("\\.m01", "", name)) %>%
  left_join(gensas.genes %>% dplyr::select(seqname, id, name), by="name") 
```

```{r}
# All DMLs 
DML.ratios

# Loci that are hypo-methylated in high pCO2 treatment
DML.ratios %>%
  filter(ratio_H.L<1)

# Loci that are hyper-methylated in high pCO2 treatment
DML.ratios %>%
  filter(ratio_H.L>1)
```

Use `bedtools` to identify where DMLs intersect with genes and gene-flanking regions 

This was done outside of RStudio on Sedna using the script [bedtools-feature-files.sh](../scripts/bedtools-feature-files.sh)

```{r}
dml25 %>% head()
```

Read in results to examine genes and gene-flanking regions containing DMLs 

```{r}
## This code only reads in results for DMLs within genes 
# dml25.annot <- 
# read_delim("../results/methylkit/dml25-genes.tab", delim = "\t", col_names = c("chr", "start_dml", "end_dml", "meth.diff", names(gensas.genes))) %>%
#   dplyr::select(-seqname) %>%
#   separate(attribute, sep = ";", into = c("ID", "Name", "Parent")) %>%
#   mutate(ID=str_remove_all(ID, "ID="), Name=str_remove_all(Name, "Name="), Parent=str_remove_all(Parent, "Parent=")) %>%
#   dplyr::rename("start_gene"="start", "end_gene"="end")

# Read in results for DMLs within genes and gene-flanking regions 
dml25.annot <- vector(mode = "list", length = 3)
names(dml25.annot) <- c("in-genes", "2kb-up-genes", "2kb-down-genes")

for (i in 1:3) {
  dml25.annot[[i]] <- 
    read_delim(file = paste("../results/methylkit/dml25-", names(dml25.annot)[i], ".tab", sep=""), delim = "\t", 
               col_names = c("chr", "start_dml", "end_dml", "meth.diff", names(gensas.genes)))
}

dml25.annot <- bind_rows(dml25.annot, .id = "location")  %>%
  dplyr::select(-seqname) %>%
  separate(attribute, sep = ";", into = c("ID", "Name", "Parent")) %>%
  mutate(ID=str_remove_all(ID, "ID="), Name=str_remove_all(Name, "Name="), Parent=str_remove_all(Parent, "Parent="), location=as.factor(location)) %>%
  dplyr::rename("start_feature"="start", "end_feature"="end") 
```

```{r}
# If a DML is inside a gene, that takes precedent over DMLs that might be up/downstream of other genes. 
# Then, for DMLs that overlap with multiple up/downstream flanking regions, identify the coordinate nearest the gene 

## Here, identify DMLs that overlap with both genes and up/down which I will for sure keep only the genes
keep.feature1 <- dml25.annot %>% 
  filter(feature=="gene") %>% 
  group_by(chr, start_dml, end_dml) %>% filter(n()>1) %>%
  arrange(chr, start_dml, end_dml) %>%
  left_join(gensas.blast, by = c("ID"="id")) %>%
  #dplyr::select(ID, location, start_dml, end_dml, start_feature, end_feature, accession, e_value) %>%
  group_by(chr, start_dml, end_dml) %>%
  filter(location=="in-genes") %>%
  mutate(locus=start_dml+1) %>%
  mutate(dml=paste(chr, locus, sep = "."))
  
## Now identify which entries to remove because the DML already overlaps with a gene 
remove.feature1 <- dml25.annot %>% 
  filter(feature=="gene") %>% 
  group_by(chr, start_dml, end_dml) %>% filter(n()>1) %>%
  filter(location!="in-genes") %>%
  mutate(locus=start_dml+1) %>%
  mutate(dml=paste(chr, locus, sep = ".")) %>%
  filter(dml %in% keep.feature1$dml) %>%
  mutate(dml.feature=paste(chr,start_dml,start_feature, sep = "."))

# ## Now identify DMLs that overlap with more than one up/downstream feature, and keep one nearest the gene  
keep.feature2 <- dml25.annot %>% filter(feature=="gene") %>% 
  group_by(chr, start_dml, end_dml) %>% filter(n()>1) %>%  # find DMLs overlapping multiple features
  filter(location!="in-genes") %>% # don't keep those overlapping gene bodies
  mutate(locus=start_dml+1) %>% mutate(dml=paste(chr, locus, sep = ".")) %>% # calculate dml locus, then create dml ID from chromosome + locus 
  filter(dml %!in% keep.feature1$dml) %>% # remove all DMLs that we've already dealt with above (i.e. they overlap gene bodies + flanking regions, so keeping just gene body feature)
  arrange(dml) %>% # arrange by DML so we can see that they overlap multiple features 
  # rename("start_feature"="start_feature", "end_feature"="end_gene") %>%  # clarify that the start/stop location in current dataframe refers to flanking features 
  left_join(gensas.genes %>% dplyr::select(name, start, end) %>%  rename(start_gene=start, end_gene=end), by = c("Name"="name")) %>% #add gene start & stop 
  mutate(dist_to_gene=case_when(  #for each DML, calculate the distance from the 1) start of gene, for upstream flanking features, and 2) end of  gene, for downstream flanking features 
    location=="2kb-up-genes" ~ start_gene-locus,
    location=="2kb-down-genes" ~ locus-end_gene)) %>%
  dplyr::select(Name, start_gene, end_gene, start_feature, end_feature, locus, dist_to_gene, everything()) %>% # reorganize columns to double check that this all worked! 
  group_by(dml) %>% slice_min(dist_to_gene) %>% # for DMLs overlapping multiple flanking regions, find one closest to the gene! 
  mutate(dml.feature=paste(chr,start_dml,start_feature, sep = "."))

## Now identify which entries to remove because the DML overlaps with >1 flanking region (remove feature with DML furthest from the gene) 
remove.feature2 <- dml25.annot %>% filter(feature=="gene") %>% 
  group_by(chr, start_dml, end_dml) %>% filter(n()>1) %>%  # find DMLs overlapping multiple features
  filter(location!="in-genes") %>% # don't keep those overlapping gene bodies
  mutate(locus=start_dml+1) %>% mutate(dml=paste(chr, locus, sep = ".")) %>% # calculate dml locus, then create dml ID from chromosome + locus 
  filter(dml %!in% keep.feature1$dml) %>% # remove all DMLs that we've already dealt with above (i.e. they overlap gene bodies + flanking regions, so keeping just gene body feature)
  arrange(dml) %>% # arrange by DML so we can see that they overlap multiple features 
  # rename("start_feature"="start_feature", "end_feature"="end_gene") %>%  # clarify that the start/stop location in current dataframe refers to flanking features 
  left_join(gensas.genes %>% dplyr::select(name, start, end) %>%  rename(start_gene=start, end_gene=end), by = c("Name"="name")) %>% #add gene start & stop 
  mutate(dist_to_gene=case_when(  #for each DML, calculate the distance from the 1) start of gene, for upstream flanking features, and 2) end of  gene, for downstream flanking features 
    location=="2kb-up-genes" ~ start_gene-locus,
    location=="2kb-down-genes" ~ locus-end_gene)) %>%
  dplyr::select(Name, start_gene, end_gene, start_feature, end_feature, locus, dist_to_gene, everything()) %>% # reorganize columns to double check that this all worked! 
  mutate(dml.feature=paste(chr,start_dml,start_feature, sep = ".")) %>%
  filter(dml.feature %!in% keep.feature2$dml.feature)
```

FINAL dataframe with gene bodies & flanking regions containing DMLs, NO DUPLICATE DMLs

```{r}
(dml25.annot.filt <- dml25.annot %>% #filter(feature=="gene") %>% 
  mutate(dml.feature=paste(chr,start_dml,start_feature, sep = ".")) %>%
  filter(dml.feature %!in% c(remove.feature1$dml.feature, remove.feature2$dml.feature)) %>%
 left_join(gensas.blast, by = c("Name"="name"))) #%>%

  # group_by(chr, start_dml, end_dml) %>% filter(n()>1) %>%  # find DMLs overlapping multiple features
  # mutate(locus=start_dml+1) %>% mutate(dml=paste(chr, locus, sep = ".")) %>% # calculate dml locus, then create dml ID from chromosome + locus
  # arrange(dml)
```




How many DMLs overlap with features? 

```{r}
paste("DML's total: ", dml25.annot.filt %>% distinct(chr, start_dml, end_dml, .keep_all = T)  %>% nrow()) 
paste("DML's in genes: ", dml25.annot.filt %>% filter(feature=="gene" & location=="in-genes") %>% distinct(chr, start_dml, end_dml, .keep_all = T)  %>% nrow()) 
paste("DML's in mRNA: ", dml25.annot.filt %>% filter(feature=="mRNA") %>% distinct(chr, start_dml, end_dml, .keep_all = T)  %>% nrow())
paste("DML's in exons: ", dml25.annot.filt %>% filter(feature=="exon") %>% distinct(chr, start_dml, end_dml, .keep_all = T) %>% nrow())
paste("DML's in CDS: ", dml25.annot.filt %>% filter(feature=="CDS") %>% distinct(chr, start_dml, end_dml, .keep_all = T)  %>% nrow()) #exon/CDS same 
paste("DML's 2kb upstream of genes: ", dml25.annot.filt %>% filter(feature=="gene" & location=="2kb-up-genes") %>% distinct(chr, start_dml, end_dml, .keep_all = T)  %>% nrow()) 
paste("DML's 2kb downstream of genes: ", dml25.annot.filt %>% filter(feature=="gene" & location=="2kb-down-genes") %>% distinct(chr, start_dml, end_dml, .keep_all = T)  %>% nrow()) 
```
Read in results to examine genes and gene-flanking regions containing all examined CpG loci  

```{r}
all.meth.annot <- vector(mode = "list", length = 3)
names(all.meth.annot) <- c("in-genes", "2kb-up-genes", "2kb-down-genes")

for (i in 1:3) {
  all.meth.annot[[i]] <- 
    read_delim(file = paste("../results/methylkit/all-mbd-", names(all.meth.annot)[i], ".tab", sep=""), delim = "\t", 
               col_names = c("chr", "start_dml", "end_dml", "meth.diff", names(gensas.genes)))
}

all.meth.annot <- bind_rows(all.meth.annot, .id = "location")  %>%
  dplyr::select(-seqname) %>%
  separate(attribute, sep = ";", into = c("ID", "Name", "Parent")) %>%
  mutate(ID=str_remove_all(ID, "ID="), Name=str_remove_all(Name, "Name="), Parent=str_remove_all(Parent, "Parent="), location=as.factor(location)) %>%
  dplyr::rename("start_feature"="start", "end_feature"="end")
```
```{r}
# If a DML is inside a gene, that takes precedent over DMLs that might be up/downstream of other genes. 
# Then, for DMLs that overlap with multiple up/downstream flanking regions, identify the coordinate nearest the gene 

## Here, identify DMLs that overlap with both genes and up/down which I will for sure keep only the genes
keep.feature1.all <- all.meth.annot %>% 
  filter(feature=="gene") %>% 
  group_by(chr, start_dml, end_dml) %>% filter(n()>1) %>%
  arrange(chr, start_dml, end_dml) %>%
  left_join(gensas.blast, by = c("ID"="id")) %>%
  #dplyr::select(ID, location, start_dml, end_dml, start_feature, end_feature, accession, e_value) %>%
  group_by(chr, start_dml, end_dml) %>%
  filter(location=="in-genes") %>%
  mutate(locus=start_dml+1) %>%
  mutate(dml=paste(chr, locus, sep = "."))
  
## Now identify which entries to remove because the DML already overlaps with a gene 
remove.feature1.all <- all.meth.annot %>% 
  filter(feature=="gene") %>% 
  group_by(chr, start_dml, end_dml) %>% filter(n()>1) %>%
  filter(location!="in-genes") %>%
  mutate(locus=start_dml+1) %>%
  mutate(dml=paste(chr, locus, sep = ".")) %>%
  filter(dml %in% keep.feature1$dml) %>%
  mutate(dml.feature=paste(chr,start_dml,start_feature, sep = "."))

# ## Now identify DMLs that overlap with more than one up/downstream feature, and keep one nearest the gene  
keep.feature2.all <- all.meth.annot %>% filter(feature=="gene") %>% 
  group_by(chr, start_dml, end_dml) %>% filter(n()>1) %>%  # find DMLs overlapping multiple features
  filter(location!="in-genes") %>% # don't keep those overlapping gene bodies
  mutate(locus=start_dml+1) %>% mutate(dml=paste(chr, locus, sep = ".")) %>% # calculate dml locus, then create dml ID from chromosome + locus 
  filter(dml %!in% keep.feature1$dml) %>% # remove all DMLs that we've already dealt with above (i.e. they overlap gene bodies + flanking regions, so keeping just gene body feature)
  arrange(dml) %>% # arrange by DML so we can see that they overlap multiple features 
  # rename("start_feature"="start_feature", "end_feature"="end_gene") %>%  # clarify that the start/stop location in current dataframe refers to flanking features 
  left_join(gensas.genes %>% dplyr::select(name, start, end) %>%  rename(start_gene=start, end_gene=end), by = c("Name"="name")) %>% #add gene start & stop 
  mutate(dist_to_gene=case_when(  #for each DML, calculate the distance from the 1) start of gene, for upstream flanking features, and 2) end of  gene, for downstream flanking features 
    location=="2kb-up-genes" ~ start_gene-locus,
    location=="2kb-down-genes" ~ locus-end_gene)) %>%
  dplyr::select(Name, start_gene, end_gene, start_feature, end_feature, locus, dist_to_gene, everything()) %>% # reorganize columns to double check that this all worked! 
  group_by(dml) %>% slice_min(dist_to_gene) %>% # for DMLs overlapping multiple flanking regions, find one closest to the gene! 
  mutate(dml.feature=paste(chr,start_dml,start_feature, sep = "."))

## Now identify which entries to remove because the DML overlaps with >1 flanking region (remove feature with DML furthest from the gene) 
remove.feature2.all <- all.meth.annot %>% filter(feature=="gene") %>% 
  group_by(chr, start_dml, end_dml) %>% filter(n()>1) %>%  # find DMLs overlapping multiple features
  filter(location!="in-genes") %>% # don't keep those overlapping gene bodies
  mutate(locus=start_dml+1) %>% mutate(dml=paste(chr, locus, sep = ".")) %>% # calculate dml locus, then create dml ID from chromosome + locus 
  filter(dml %!in% keep.feature1$dml) %>% # remove all DMLs that we've already dealt with above (i.e. they overlap gene bodies + flanking regions, so keeping just gene body feature)
  arrange(dml) %>% # arrange by DML so we can see that they overlap multiple features 
  # rename("start_feature"="start_feature", "end_feature"="end_gene") %>%  # clarify that the start/stop location in current dataframe refers to flanking features 
  left_join(gensas.genes %>% dplyr::select(name, start, end) %>%  rename(start_gene=start, end_gene=end), by = c("Name"="name")) %>% #add gene start & stop 
  mutate(dist_to_gene=case_when(  #for each DML, calculate the distance from the 1) start of gene, for upstream flanking features, and 2) end of  gene, for downstream flanking features 
    location=="2kb-up-genes" ~ start_gene-locus,
    location=="2kb-down-genes" ~ locus-end_gene)) %>%
  dplyr::select(Name, start_gene, end_gene, start_feature, end_feature, locus, dist_to_gene, everything()) %>% # reorganize columns to double check that this all worked! 
  mutate(dml.feature=paste(chr,start_dml,start_feature, sep = ".")) %>%
  filter(dml.feature %!in% keep.feature2$dml.feature)

# FINAL dataframe with gene bodies & flanking regions containing all examined CpG loci, NO DUPLICATE CPGS

all.meth.annot.filt <- all.meth.annot %>% #filter(feature=="gene") %>% 
  mutate(dml.feature=paste(chr,start_dml,start_feature, sep = ".")) %>%
  filter(dml.feature %!in% c(remove.feature1.all$dml.feature, remove.feature2.all$dml.feature)) %>% 
 left_join(gensas.blast, by = c("Name"="name")) #%>%

  # group_by(chr, start_dml, end_dml) %>% filter(n()>1) %>%  # find DMLs overlapping multiple features
  # mutate(locus=start_dml+1) %>% mutate(dml=paste(chr, locus, sep = ".")) %>% # calculate dml locus, then create dml ID from chromosome + locus
  # arrange(dml)
```



```{r}
paste("CpG's total: ", all.meth.annot %>% distinct(chr, start_dml, end_dml, .keep_all = T)  %>% nrow()) 
paste("CpG's in genes: ", all.meth.annot %>% filter(feature=="gene" & location=="in-genes") %>% distinct(chr, start_dml, end_dml, .keep_all = T)  %>% nrow()) 
paste("CpG's in mRNA: ", all.meth.annot %>% filter(feature=="mRNA") %>% distinct(chr, start_dml, end_dml, .keep_all = T)  %>% nrow())
paste("CpG's in exons: ", all.meth.annot %>% filter(feature=="exon") %>% distinct(chr, start_dml, end_dml, .keep_all = T) %>% nrow())
paste("CpG's in CDS: ", all.meth.annot %>% filter(feature=="CDS") %>% distinct(chr, start_dml, end_dml, .keep_all = T)  %>% nrow()) #exon/CDS same 
paste("CpG's 2kb upstream of genes: ", all.meth.annot %>% filter(feature=="gene" & location=="2kb-up-genes") %>% distinct(chr, start_dml, end_dml, .keep_all = T)  %>% nrow()) 
paste("CpG's 2kb downstream of genes: ", all.meth.annot %>% filter(feature=="gene" & location=="2kb-down-genes") %>% distinct(chr, start_dml, end_dml, .keep_all = T)  %>% nrow()) 
```
Summarize DML and CpG locations along gene features

```{r}
(features <- dml25.annot.filt %>% group_by(location, feature) %>%
  summarize(DMLs=n()) %>% 
  left_join(all.meth.annot.filt %>% group_by(location, feature) %>%
  summarize(CpGs=n())) %>% ungroup() %>%
  add_row(location="total", feature="any", 
          DMLs=dml25.annot.filt %>% distinct(chr, start_dml, end_dml, .keep_all = T)  %>% nrow(), 
          CpGs=all.meth.annot.filt %>% distinct(chr, start_dml, end_dml, .keep_all = T)  %>% nrow()) %>%
  mutate(feature=case_when(
  location=="2kb-down-genes" ~ "2kb-down-genes",
  location=="2kb-up-genes" ~ "2kb-up-genes",
  TRUE ~ feature)) %>%
   mutate(DML.perc=DMLs/906, CpG.perc=CpGs/91225))
```


```{r}
meth.chisq <- features %>% filter(feature %in% c("2kb-up-genes", "2kb-down-genes", "exon", "gene")) %>%
  column_to_rownames("feature") %>% dplyr::select(CpGs, DMLs) %>% chisq.test() 

meth.chisq #difference in overall distribution
meth.chisq$observed %>% sum()
meth.chisq$expected
corrplot(meth.chisq$residuals, is.cor = FALSE, method="circle", cl.pos = 'n')
```

**Positive residuals are in blue.** Positive values in cells specify an attraction (positive association) between the corresponding row and column variables. 
--> Upstream and downstream flanking regions are more likely to be methylated. 

**Negative residuals are in red.** This implies a repulsion (negative association) between the corresponding row and column variables. 
--> exons and genes less likely 

### Examine the contribution of each genome feature to the chi-square value (in percentage) 

The up- and down-stream flanking contribute most of the chi-square value 

```{r}
contrib <- 100*meth.chisq$residuals^2/meth.chisq$statistic
round(contrib, 3) 
corrplot(contrib, is.cor = F)
```
```{r}
features %>% dplyr::select(feature, DML.perc, CpG.perc) %>%
  pivot_longer(cols = c("DML.perc", "CpG.perc"), names_to = "analysis", values_to = "percent") %>%
  filter(feature %in% c("2kb-down-genes", "2kb-up-genes", "gene", "exon")) %>%
  mutate(percent2 = ifelse(analysis == "CpGs", percent*(-1), percent)) %>%
  mutate(feature=fct_rev(feature), analysis=factor(analysis, ordered = T, levels=c("DML.perc", "CpG.perc"))) %>%

ggplot(aes(x=feature, y=percent, fill=analysis, label=percent(percent, accuracy = 0.1))) +  #prettyNum(count, big.mark = ",")
  geom_bar(stat="identity", position = "dodge", width = .5) +
scale_fill_manual(name = NULL, labels = c("CpG.perc"="All analyzed loci", "DML.perc"="Differentially methylated loci"),
                  values=c("gray40", "gray65")) +
  ggtitle("% of DML's and all analyzed loci that overlap with gene features") +
  labs(y="% of Loci", x=NULL) +
  theme_minimal() + theme(legend.position ="bottom") + 
    geom_text(aes(label=percent(percent, accuracy = 1)), 
              position=position_dodge(width=0.5), vjust=-0.5, size=3) +
  scale_x_discrete(
    labels=c(gene="Gene", exon="Exon", `2kb-up-genes`="5' flanking\nregion (-2kb)", 
             `2kb-down-genes`="3' flanking\nregion (+2kb)"))
```

How many genes contain DMLs?

```{r}
# 906 DMLs overlap with 650 distinct genes or gene-flanking regions
dml25.annot.filt %>% 
  filter(feature=="gene") %>% group_by(chr, start_dml, end_dml) %>%
  arrange(chr, start_dml, end_dml) %>%
  left_join(gensas.blast, by = c("Name"="name")) %>% ungroup() %>% 
  dplyr::select(ID) %>% distinct() %>% nrow() #comment out this line to find unique genes 

# 581 DMLs overlap with genes 
(dml25.genes <- dml25.annot.filt %>% ungroup() %>%
  filter(location=="in-genes", feature=="gene"))

# 422 distinct genes 
dml25.genes %>% ungroup() %>% dplyr::select(chr, start_feature, end_feature) %>% distinct() %>% nrow()

# 245 distinct genes are annotated with Uniprot SPID
dml25.genes %>% ungroup() %>% filter(!is.na(accession)) %>% dplyr::select(chr, start_feature, end_feature) %>% distinct() %>% nrow()
```

```{r}
# 325 DMLs overlap with gene flanking regions 
(dml25.flank.genes <- dml25.annot.filt %>% 
  filter(location %in% c("2kb-up-genes", "2kb-down-genes"), feature=="gene"))

# 271 distinct genes 
dml25.flank.genes %>% dplyr::select(chr, start_feature, end_feature) %>% distinct() %>% nrow()

# 93 genes are annotated with Uniprot SPID
dml25.flank.genes  %>% filter(!is.na(accession)) %>% dplyr::select(chr, start_feature, end_feature) %>% distinct() %>% nrow()
```
How many genes have DMLs both in the gene and flanking the gene?

```{r}
# 45 distinct genes 
dml25.genes %>% filter(ID %in% dml25.flank.genes$ID) %>% dplyr::select(ID) %>% distinct() %>% nrow()
```

NOTE: If more than one DML is located in a gene, that gene appears multiple times - once per DML 

How many genes contain more than one DML? Answer: 94 genes contain a total of 253 DMLs (out of a total of 581 DMLs)

```{r}
dml25.genes %>% group_by(chr, start_feature, end_feature) %>% mutate(dupe=n()>1) %>% filter(dupe==TRUE) %>%
  dplyr::select(chr, start_feature, end_feature) %>% distinct() %>% nrow()

dml25.genes.multi <- 
  dml25.genes %>% group_by(chr, start_feature, end_feature) %>% mutate(dupe=n()>1) %>% filter(dupe==TRUE)
```

How many gene-flanking regions contain more than one DML? Answer: 40 genes contain a total of 94 DMLs 

```{r}
dml25.flank.genes %>% group_by(chr, start_feature, end_feature) %>% mutate(dupe=n()>1) %>% filter(dupe==TRUE) %>%
  dplyr::select(chr, start_feature, end_feature) %>% distinct() %>% nrow()

dml25.flank.genes.multi <- 
  dml25.flank.genes %>% group_by(chr, start_feature, end_feature) %>% mutate(dupe=n()>1) %>% filter(dupe==TRUE)

# How many multi-DML upstream gene flanking regions? 
dml25.flank.genes.multi %>% filter(location=="2kb-up-genes") #68 DMLs upstream genes 
dml25.flank.genes.multi %>% filter(location=="2kb-up-genes") %>% dplyr::select(chr, start_feature, end_feature) %>% distinct() %>% nrow() # in 28 genes 

# How many multi-DML downstream gene flanking regions? 
dml25.flank.genes.multi %>% filter(location=="2kb-down-genes") #26 DMLs downstream genes 
dml25.flank.genes.multi %>% filter(location=="2kb-down-genes") %>% dplyr::select(chr, start_feature, end_feature) %>% distinct() %>% nrow() # in 12 genes 
```
How many genes contain more than one DML in the greater gene area? 

347 DMLs overlap with 134 genes that contain multiple DMLs 

```{r}
dml25.annot.filt %>% filter(feature=="gene") %>% group_by(chr, start_feature, end_feature) %>% 
  mutate(dupe=n()>1) %>% filter(dupe==TRUE) %>% # 347 DMLs 
  dplyr::select(chr, start_feature, end_feature) %>% distinct() %>% nrow() # in 134 genes 

dml25.gene.regions.multi <- 
  dml25.annot.filt %>% filter(feature=="gene") %>% group_by(chr, start_feature, end_feature) %>% mutate(dupe=n()>1) %>% filter(dupe==TRUE)
```

Which gene regions (bodies + up/down) contain more than one DML? What do they do?

```{r}
dml25.annot.filt %>% filter(feature=="gene") %>% group_by(Name, chr, start_feature, end_feature) %>% tally() %>%
  arrange(desc(n)) %>% filter(n>1) %>%
  left_join(dml25.annot.filt %>% filter(feature=="gene") %>% dplyr::select(Name, chr, start_feature, end_feature, accession, match_id, e_value)) %>%
  distinct() %>% arrange(chr, start_feature)
```

Which gene bodies contain more than one DML? What do they do?

```{r}
dml25.genes %>% group_by(Name, chr, start_feature, end_feature) %>% tally() %>%
  arrange(desc(n)) %>% filter(n>1) %>%
  left_join(dml25.genes %>% dplyr::select(Name, chr, start_feature, end_feature, accession, match_id, e_value)) %>%
  distinct() %>% arrange(chr, start_feature)
```

Which upstream gene regions contain more than one DML? What do they do?

```{r}
dml25.flank.genes.up.annot <- dml25.flank.genes %>% filter(location=="2kb-up-genes") %>% 
  group_by(chr, start_feature, end_feature) %>% tally() %>%
  arrange(desc(n)) %>% filter(n>1) %>%
  left_join(dml25.flank.genes %>% dplyr::select(chr, start_feature, end_feature, accession, match_id, e_value, id)) %>%
  distinct() %>% ungroup() %>%
  dplyr::select(accession) %>%  na.omit() %>% unlist() %>% as.vector() #%>% write_clip()
```
Which downstream gene regions contain more than one DML? What do they do?

```{r}
dml25.flank.genes.down.annot <- dml25.flank.genes %>% filter(location=="2kb-down-genes") %>% 
  group_by(chr, start_feature, end_feature) %>% tally() %>%
  arrange(desc(n)) %>% filter(n>1) %>%
  left_join(dml25.flank.genes %>% dplyr::select(chr, start_feature, end_feature, accession, match_id, e_value, id)) %>%
  distinct() %>% ungroup() %>%
  dplyr::select(accession) %>%  na.omit() %>% unlist() %>% as.vector() #%>% write_clip()
```
## Examine function of genes that contain more than 3 DMLs 

Export Uniprot IDs for annotated genes with 2 or more DMLs and see what they do 

414 DMLs within 158 genes, 77 with Uniprot IDs

```{r}
dml25.annot.filt %>% filter(feature=="gene") %>% group_by(Name) %>% tally() %>%
  arrange(desc(n)) %>% filter(n>1) %>%
  left_join(dml25.annot.filt %>% filter(feature=="gene") %>% dplyr::select(Name, accession, match_id, e_value)) %>%
  distinct() %>% filter(n>1) %>%
  filter(!is.na(accession)) %>% ungroup() %>%
  dplyr::select(accession) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```

Export Uniprot IDs for annotated genes with 3 or more DMLs and see what they do 
188 DMLs within 45 genes, 21 that are annotated

```{r}
dml25.annot.filt %>% filter(feature=="gene") %>% group_by(Name) %>% tally() %>%
  arrange(desc(n)) %>% filter(n>1) %>%
  left_join(dml25.annot.filt %>% filter(feature=="gene") %>% dplyr::select(Name, accession, match_id, e_value)) %>%
  distinct() %>% filter(n>2) %>%
  filter(!is.na(accession)) %>% ungroup() %>%
  dplyr::select(accession) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```

Pasted the above list of Uniprot IDs into the Uniprot batch retrieval tool, and downloaded. Import and view results 

```{r}
# Import Uniprot details for genes containing multiple DMLs and generate wordclouds

#file <- "uniprot-multi-dml-genes_2plus.tsv" #for 2+ DMLs per gene 
file <- "uniprot-multi-dml-genes_3plus.tsv" #for 3+ DMLs per gene 

# Word cloud with Uniprot Keywords
read_delim(file = paste("../results/methylkit/", file, sep=""), delim = "\t") %>% clean_names() %>%
  dplyr::select(keywords) %>% na.omit() %>% unlist() %>% as.vector() %>% 
#  gsub("\\[[^\\]]*\\]", "", ., perl=TRUE) %>%
#  gsub("process", "", ., perl=TRUE) %>% #get rid of word "process" since it's in everything
  wordcloud(min.freq = 2, random.order = F, random.color = F, colors = brewer.pal(2, "Dark2"))

# Word cloud with GO Biological Processes
tiff(filename = "../results/methylkit/wordcloud-3DML-GO-BP.tiff", width = 9, height = 9, units = "in", res=500)
read_delim(file = paste("../results/methylkit/", file, sep=""), delim = "\t") %>% clean_names() %>%
  dplyr::select(gene_ontology_biological_process) %>% na.omit() %>% unlist() %>% as.vector() %>% 
  gsub("\\[[^\\]]*\\]", "", ., perl=TRUE) %>%
  gsub("process", "", ., perl=TRUE) %>% #get rid of word "process" since it's in everything
  wordcloud(min.freq = 2, random.order = F, random.color = F, colors = brewer.pal(2, "Dark2"))
dev.off()

# Word cloud with GO Molecular Function
read_delim(file = paste("../results/methylkit/", file, sep=""), delim = "\t") %>% clean_names() %>%
  dplyr::select(gene_ontology_molecular_function) %>% na.omit() %>% unlist() %>% as.vector() %>% 
  gsub("\\[[^\\]]*\\]", "", ., perl=TRUE) %>%
  gsub("process", "", ., perl=TRUE) %>% #get rid of word "process" since it's in everything
  wordcloud(min.freq = 2, random.order = F, random.color = F, colors = brewer.pal(2, "Dark2"))

# Word cloud with GO Cellular Componenet
read_delim(file = paste("../results/methylkit/", file, sep=""), delim = "\t") %>% clean_names() %>%
  dplyr::select(gene_ontology_cellular_component) %>% na.omit() %>% unlist() %>% as.vector() %>% 
  gsub("\\[[^\\]]*\\]", "", ., perl=TRUE) %>%
  gsub("process", "", ., perl=TRUE) %>% #get rid of word "process" since it's in everything
  wordcloud(min.freq = 2, random.order = F, random.color = F, colors = brewer.pal(2, "Dark2"))
```

Heatmap with DMLs that are contained within multi-DML genes 

```{r}
#ggplotly(
ggplot(
  # Filter DMLs for those that are contained within multi-DML genes 
  DMLs_heatmap %>% mutate(locus=paste(chr, ".", start, sep = "")) %>% 
  filter(locus %in% 
           (dml25.gene.regions.multi %>% mutate(locus=paste(chr, ".", start_dml+1, sep = "")))$locus) %>%
    # Sort loci by the ratio of mean meth in H:L CO2 (groups hypo- and hyper- methylated loci together)
    left_join(DML.ratios %>% mutate(locus=paste(chr, ".", start, sep = "")) %>% dplyr::select(locus, ratio_H.L),
              by="locus") %>% 
    arrange(ratio_H.L) %>% 
    # lock in locus order 
    mutate(chr_start=factor(chr_start, ordered = T)), 
  
  aes(sample, chr_start, fill= percMeth)) + 
  xlab("Sample") + geom_tile(na.rm = T) +
  #scale_y_discrete(limits=(DML.ratios[order(DML.ratios$ratio_H.L),]$chr_start)) + 
  #scale_fill_viridis(discrete=FALSE) 
  theme(axis.title.y=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank()) + 
  scale_fill_distiller(palette = "YlGnBu", direction = 1) #)
  #scale_fill_gradient(low="gray5", high="white")
```
```{r}
DML.ratios
```

Create figure showing multi-DML genes

```{r}
test <- 
#  dml25.genes %>%  filter(Name %in% dml25.genes.multi$Name) %>%  right_join(dml25.genes %>%  #use DMLs in gene bodies only 
  dml25.annot.filt %>% filter(feature=="gene") %>%  filter(Name %in% dml25.gene.regions.multi$Name) %>% right_join(dml25.annot.filt %>% filter(feature=="gene") %>% #use DMLs in gene bodies + flanking regions  
    group_by(Name) %>% tally() %>% arrange(desc(n)) %>% 
               filter(n>1)) %>% #here you can filter for the number of DMLs in a gene
  arrange(n) %>% mutate(Name=factor(Name, ordered = T)) %>%
  mutate(n=factor(n, ordered = T)) %>% 
  group_by(Name) %>% summarize(mean=mean(-1*meth.diff), sd=sd(-1*meth.diff))

# holy crab this function is amazing! 
#forcats:: fct_inorder()

# IMPORTANT! meth.diff < 0 = loci where methylation rate is HIGHER in high pCO2 treatment 
# To make this intuitive in the below figure, we must multiply meth.diff by -1 so negative numbers = less methylated in treatment 

ggplotly(
#  dml25.genes %>% filter(Name %in% dml25.genes.multi$Name) %>% right_join(dml25.genes %>%  #use DMLs in gene bodies only 
  dml25.annot.filt %>% filter(feature=="gene") %>% filter(Name %in% dml25.gene.regions.multi$Name) %>% right_join(dml25.annot.filt %>% filter(feature=="gene") %>%   #use DMLs in gene bodies + flanking regions
                                      group_by(Name) %>% tally() %>% arrange(desc(n)) %>% 
               filter(n>1)) %>% #here you can filter for the number of DMLs in a gene
  left_join(test) %>% filter(!is.na(location)) %>% droplevels("location") %>%
  arrange(n, mean) %>% mutate(Name=factor(Name, ordered = T)) %>%
  mutate(n=factor(n, ordered = T)) %>%
  
  ggplot(aes(x=fct_inorder(Name), y=-1*meth.diff, color=n, shape=fct_rev(location), label=accession)) + 
    geom_point(size=4, alpha=0.6) +
  coord_flip() + theme_minimal() + 
  scale_colour_brewer(palette = "Dark2") + 
  theme(axis.text.y = element_blank()) + 
  guides(colour = guide_legend(reverse=T, title="Number of\nDMLs in gene", order = 1), 
         shape = guide_legend(title="Location of DML")) + 
  scale_shape_discrete(labels=c("in-genes"="Gene bodies", "2kb-up-genes"="5' (upstream)", "2kb-down-genes"="3' (downstream)")) + 
  xlab("Genes") + 
  ylab("Methylation difference (%)\nbetween pCO2 treatments") +
  stat_summary(fun.data = "mean_sdl", fun.args=list(mult=1), geom="crossbar", width=0.2), 
  tooltip=c("x", "y", "label") 
  #ggplot2::annotate("text", x = c(60, 52), y=c(-55, 60), label = c("Lower methylation in\nOA crab", "Higher methylation in\nOA crab"), size=3, color="gray40")
)
```


```{r}
# all.annot <- 
# read_delim("../results/methylkit/all-mbd-genes.tab", delim = "\t", col_names = c("chr", "start_dml", "end_dml", "meth.diff", names(gensas.genes))) %>%
#   dplyr::select(-seqname) %>%
#   separate(attribute, sep = ";", into = c("ID", "Name", "Parent")) %>%
#   mutate(ID=str_remove_all(ID, "ID="), Name=str_remove_all(Name, "Name="), Parent=str_remove_all(Parent, "Parent=")) %>%
#   rename("start"="start_feature", "end"="end_feature")
# 
# # 60,199 genes contain an MBD locus that we analyzed 
# all.genes <- all.annot %>% filter(feature=="gene") %>%
#   left_join(gensas.blast, by = c("Name"="name"))

# 84,433 CpG loci 
all.cpg <- all.meth.annot.filt %>% filter(feature=="gene") %>% 
  rename("start_cpg"="start_dml", "end_cpg"="end_dml")

# Aligned to 4,539 genes
all.genes <- all.cpg %>% dplyr::select(Name, accession) %>% distinct()

# 1,969 of those are annotated with a SPID
all.genes %>% filter(!is.na(accession)) 
```

DAVID Enrichment analysis

```{r}
DMLs.4David <- list(
  
  # All DMLs that map to Uniprot SPIDs
  "all-dmls" = 
      dml25.annot.filt %>% filter(feature=="gene") %>%  
      dplyr::select(Name, accession) %>% distinct() %>%
      dplyr::select(accession) %>% filter(!is.na(accession)) %>%
      na.omit() %>% unlist() %>% as.vector(),
  
  # Gene-body DMLs
  "gene-bodies" = 
      dml25.annot.filt %>% filter(feature=="gene") %>%  
      filter(location=="in-genes") %>% 
      dplyr::select(Name, accession) %>% distinct() %>%
      dplyr::select(accession) %>% filter(!is.na(accession)) %>%
      na.omit() %>% unlist() %>% as.vector(),

  # 5' "upstream" DMLs
  "upstream" = 
      dml25.annot.filt %>% filter(feature=="gene") %>%  
      filter(location=="2kb-up-genes") %>% 
      dplyr::select(Name, accession) %>% distinct() %>%
      dplyr::select(accession) %>% filter(!is.na(accession)) %>%
      na.omit() %>% unlist() %>% as.vector(),

  # 3' "downstream" DMLs
  "downstream" = 
      dml25.annot.filt %>% filter(feature=="gene") %>%  
      filter(location=="2kb-down-genes") %>% 
      dplyr::select(Name, accession) %>% distinct() %>%
      dplyr::select(accession) %>% filter(!is.na(accession)) %>%
      na.omit() %>% unlist() %>% as.vector(),

  # Gene regions with 2+ DMLs 
  "multi_2plus" = 
    dml25.annot.filt %>% filter(feature=="gene") %>%
    group_by(Name) %>% tally() %>%
    arrange(desc(n)) %>%
    left_join(dml25.annot.filt %>% filter(feature=="gene") %>% dplyr::select(Name, accession, match_id, e_value)) %>%
    distinct() %>% 
    filter(n>1) %>% # use genes with 2+ DMLs
    filter(!is.na(accession)) %>% ungroup() %>%
    dplyr::select(accession) %>%  na.omit() %>% unlist() %>% as.vector(),

    # Gene regions with 3+ DMLs 
    "multi_3plus" = 
      dml25.annot.filt %>% filter(feature=="gene") %>%
      group_by(Name) %>% tally() %>%
      arrange(desc(n)) %>%
      left_join(dml25.annot.filt %>% filter(feature=="gene") %>% dplyr::select(Name, accession, match_id, e_value)) %>%
      distinct() %>% 
      filter(n>2) %>% # use genes with 3+ DMLs
      filter(!is.na(accession)) %>% ungroup() %>%
      dplyr::select(accession) %>%  na.omit() %>% unlist() %>% as.vector()

  ) 

# Write tab-delimited file with each column containing Uniprot IDs for each set of DEGs 
write_delim(x = plyr::ldply(DMLs.4David, rbind) %>% t() %>% as.data.frame(), delim = "\t", col_names = F, na="", 
            file = "../results/methylkit/DMLs-4David.txt")


# All Analyzed loci 
all.genes %>% dplyr::select(Name, accession) %>% distinct() %>%
  filter(!is.na(accession)) %>%
  dplyr::select(accession) %>% unlist() %>% as.vector() %>% write_clip()
```

Import DAVID results 

```{r}
filenames <- read_delim(file="../results/methylkit/DAVID_filenames.txt", delim = "\t", col_names = c("filename", "gene_set"))
file_list <- vector(mode = "list", length = nrow(filenames))
names(file_list) <- c(filenames$gene_set)

for (i in 1:nrow(filenames)) {
    file_list[[i]] <- data.frame(read_delim(file=paste("../results/methylkit/", filenames[i,"filename"], sep=""), delim = "\t"))}

david <- bind_rows(file_list, .id = "gene_set") %>% clean_names() %>%  # Bind dataframe for each gene_set together
    #filter(category=="GOTERM_BP_DIRECT") %>%
    #filter(p_value<0.05) %>%
    dplyr::rename(percent=x) %>%
    separate(term, into = c("term", "process"), sep = "~") %>%
    mutate(gene_set=factor(gene_set)) %>% 
  tidyr::complete(gene_set, category) # add rows for gene_sets that are missing from dataframe
```

```{r}

david %>% 
#  filter(gene_set != "multi-dmls_3plus") %>%
  #filter(count>=3) %>% 
#  filter(p_value<0.05) %>%
#  filter(category=="GOTERM_BP_DIRECT") %>% 
#  filter(category=="GOTERM_MF_DIRECT") %>% 
#  filter(category=="GOTERM_CC_DIRECT") %>% 
  filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>%

  group_by(gene_set, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(
    david %>% filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>% 
              dplyr::select(gene_set, category, term, process, count, percent, p_value, fold_enrichment, fdr, genes), 
            by = c("gene_set", "genes", "p_value", "count")) %>%  #re-add data
#  ungroup() %>% arrange(stress, response, p_value) %>% group_by(stress, response) %>% dplyr::slice(1:15) %>% # to reduce dataset if too large

  ungroup() %>% arrange(gene_set, p_value) %>%
  mutate(gene_set=factor(gene_set, ordered = T, 
                         levels=c("all-dmls", "multi-dmls_2plus", "multi-dmls_3plus", "gene-bodies", "upstream", "downstream"))) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%

ggplot(aes(y = process, x=gene_set, col=gene_set)) + 
  geom_point(alpha=0.65, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
 scale_color_manual(name="Gene Set",
                     values=c("all-dmls"="#1b9e77", "gene-bodies"="#d95f02", "multi-dmls_2plus"="#7570b3", 
                              "multi-dmls_3plus"="#e7298a", "upstream"="#66a61e", "downstream"="#e6ab02"),
                    labels=c("all-dmls"="Genes with ≥1 DMLs", 
                             "gene-bodies"="Gene-body DMLs", 
                             "multi-dmls_2plus"="Genes with ≥2 DMLs", 
                             "multi-dmls_3plus"="Genes with ≥3 DMLs", 
                             "upstream"="5'-region DMLs", 
                             "downstream"="3'-region DMLs"),
             guide = guide_legend(override.aes = list(size=5))) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.title.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=7),
        legend.text=element_text(size=7), 
        legend.title = element_text(size=7),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
#  ggtitle("Enriched Biological Processes")
#  ggtitle("Enriched Molecular Functions")
#  ggtitle("Enriched Cellular Componenets")
  ggtitle("Enriched Biological Processes")

```

Export Uniprot IDs for annotated genes with multiple flanking DMLs and see what they do 

```{r}
dml25.flank.genes %>% group_by(ID, location) %>% tally() %>%
  arrange(desc(n)) %>% filter(n>1) %>%
  left_join(dml25.flank.genes %>% dplyr::select(ID, accession, match_id, e_value, id)) %>%
  distinct() %>% filter(n>1) %>%
  filter(!is.na(accession)) %>% ungroup() %>%
  dplyr::select(accession) %>%  na.omit() %>% unlist() %>% as.vector() %>% write_clip()
```
















### Alignment to mitochondrial DNA to estimate conversion rate 

```{r}
# Join sample info with bismark summary report 

(sample.info.mito <- sample.info %>% 
   dplyr::select(sample, tank, treatment, high_or_low_co2, sample_seq, developmental_stage, 
                 dna_siolation_batch, perc_totalread_unique, CpGs_Meth, CHGs_Meth, CHHs_Meth) %>% 
  
  right_join(

  # Read in bismark summary report
    read_delim(file="../results/bismark/mito/bismark_summary_report.txt", delim="\t") %>% mutate(File = str_replace_all(File, '_R1_bismark_bt2_pe.bam', "")) %>%
  clean_names() %>% 
   separate(file, into = c("a", "b", "sample_seq"), remove = F, sep = "_") %>%
   mutate(sample_prep=paste(a, b, sep = "-")) %>%
    mutate(sample_seq=as.numeric(str_replace_all(sample_seq, "S", ""))) %>%
   #dplyr::select(file, sample_prep, sample_seq, everything(), -a, -b) %>%
  mutate(perc_totalread_unique_mito=unique_reads_remaining/total_reads, 
         CpGs_Meth_mito=100*methylated_cp_gs/(methylated_cp_gs+unmethylated_cp_gs), 
         CHGs_Meth_mito=100*methylated_chgs/(methylated_chgs+unmethylated_chgs), 
         CHHs_Meth_mito=100*methylated_ch_hs/(methylated_ch_hs+unmethylated_ch_hs)) %>% 
    dplyr::select(sample_prep, perc_totalread_unique_mito, CpGs_Meth_mito, CHGs_Meth_mito, CHHs_Meth_mito),

  by = c("sample"="sample_prep")) %>%
  mutate(included=case_when(
    sample_seq== 1 ~ "yes",
    sample_seq== 2 ~ "yes",
    sample_seq== 3 ~ "yes",     
    sample_seq== 4 ~ "yes",
    sample_seq== 7 ~ "yes",
    sample_seq== 12 ~ "yes",
    sample_seq== 10 ~ "yes",
    sample_seq== 11 ~ "yes",
    sample_seq== 15 ~ "yes",
    sample_seq== 17 ~ "yes",
    sample_seq== 18 ~ "yes",
    sample_seq== 20 ~ "yes",
    TRUE ~ "no")))


# **Good samples:** all have mean % methylation >= 70% and nloci meeting 10x > 3e5
# control pCO2: 1, 2, 3, 4, 7, 12
# high pCO2:    10, 11, 15, 17, 18, 20

ggplot(sample.info.mito, aes(x=CpGs_Meth_mito, y=CpGs_Meth)) + 
  geom_point(aes(shape=high_or_low_co2, color=included), size=3) + 
  scale_color_manual(values=c("red", "blue")) +
           #geom_text(aes(label=sample_seq, color=high_or_low_co2), vjust = -0.7) + 
  theme_minimal() + geom_smooth(method = "lm", se=T, color="black", linetype="dotted", fill="gray80") +
  #annotate("text", x = 2.25, y = 70, label = "r = -0.38", size=6) + 
ggtitle("% CpG Methylation full genome ~ mitochondria") + 
  xlab("Mitochindria CpG meth %") + ylab("Full Genome CpG % meth")

ggplot(sample.info.mito, aes(x=perc_totalread_unique_mito, y=CpGs_Meth_mito)) + 
  geom_point(aes(shape=high_or_low_co2, color=included), size=3) + 
  scale_color_manual(values=c("red", "blue")) +
           #geom_text(aes(label=sample_seq, color=high_or_low_co2), vjust = -0.7) + 
  theme_minimal() + geom_smooth(method = "lm", se=T, color="black", linetype="dotted", fill="gray80") +
  #annotate("text", x = 2.25, y = 70, label = "r = -0.38", size=6) + 
ggtitle("mitochondria % CpG methylation ~ alignment rate") + 
  xlab("Mitochindria alignment rate") + ylab("Mitochindria CpG meth %")

ggplot(sample.info.mito, aes(x=included, y=CpGs_Meth_mito, color=included)) + 
  geom_boxplot(outlier.shape = NA) + geom_jitter(width = 0.15, aes(shape=high_or_low_co2)) + 
  theme_minimal() + scale_color_manual(values=c("red", "blue")) #+
           #geom_text(aes(label=sample_seq, color=high_or_low_co2), vjust = -0.7) + 
  #theme_minimal() + geom_smooth(method = "lm", se=T, color="black", linetype="dotted", fill="gray80") +
  #annotate("text", x = 2.25, y = 70, label = "r = -0.38", size=6) + 
#ggtitle("% CpG Methylation full genome ~ mitochondria") + 
#  xlab("Mitochindria CpG meth %") + ylab("Full Genome CpG % meth")

summary(sample.info.mito$CpGs_Meth_mito)
```

Create table of alignment rate and methylation rates in full genome and mitochondrial contexts to assess sample integrity 

```{r}
sample.info %>% dplyr::select(sample_seq, sample, high_or_low_co2, treatment, tank, developmental_stage, 
                              dna_siolation_batch, perc_totalread_unique, CpGs_Meth, CHGs_Meth, CHHs_Meth) %>%
  rename("perc_totalread_unique"="perc_totalread_unique_genome", 
         "CpGs_Meth"="CpGs_Meth_genome", "CHGs_Meth"="CHGs_Meth_genome", "CHHs_Meth"="CHHs_Meth_genome") %>% 
  mutate(perc_totalread_unique_genome= perc_totalread_unique_genome*100)  %>% 
  right_join(

  # Read in bismark summary report
    read_delim(file="../results/bismark/mito/bismark_summary_report.txt", delim="\t") %>% mutate(File = str_replace_all(File, '_R1_bismark_bt2_pe.bam', "")) %>%
  clean_names() %>% 
   separate(file, into = c("a", "b", "sample_seq"), remove = F, sep = "_") %>%
   mutate(sample_prep=paste(a, b, sep = "-")) %>%
    mutate(sample_seq=as.numeric(str_replace_all(sample_seq, "S", ""))) %>%
   #dplyr::select(file, sample_prep, sample_seq, everything(), -a, -b) %>%
  mutate(perc_totalread_unique_mito=100*unique_reads_remaining/total_reads, 
         CpGs_Meth_mito=100*methylated_cp_gs/(methylated_cp_gs+unmethylated_cp_gs), 
         CHGs_Meth_mito=100*methylated_chgs/(methylated_chgs+unmethylated_chgs), 
         CHHs_Meth_mito=100*methylated_ch_hs/(methylated_ch_hs+unmethylated_ch_hs)) %>% 
    dplyr::select(sample_prep, perc_totalread_unique_mito, CpGs_Meth_mito, CHGs_Meth_mito, CHHs_Meth_mito),

  by = c("sample"="sample_prep")) %>%
  mutate_at(vars(perc_totalread_unique_genome, CpGs_Meth_genome, CHGs_Meth_genome, CHHs_Meth_genome,
                 CpGs_Meth_mito, CHGs_Meth_mito, CHHs_Meth_mito), funs(round(., 1))) %>%
  mutate(perc_totalread_unique_mito=round(perc_totalread_unique_mito, digits = 2)) %>%
  mutate(included=case_when(
    sample_seq== 1 ~ "yes",
    sample_seq== 2 ~ "yes",
    sample_seq== 3 ~ "yes",     
    sample_seq== 4 ~ "yes",
    sample_seq== 7 ~ "yes",
    sample_seq== 12 ~ "yes",
    sample_seq== 10 ~ "yes",
    sample_seq== 11 ~ "yes",
    sample_seq== 15 ~ "yes",
    sample_seq== 17 ~ "yes",
    sample_seq== 18 ~ "yes",
    sample_seq== 20 ~ "yes",
    TRUE ~ "no")) %>%
  arrange(desc(included), desc(high_or_low_co2), sample_seq) %>% 
  write.csv(file="../results/alignment-meth-summary.csv", row.names=FALSE) 

```




## BONEYARD 

Divergent Bar plot of DMLs 

```{r}
# DML.ratios %>% 
#            mutate(hypermethylated=ifelse(diff_H.L>0, "H", "L")) %>%
#   ggplot(aes(x = chr_start, y = diff_H.L, fill=hypermethylated)) +
#   geom_bar(stat = "identity", width=0.41) +
#     scale_x_discrete(limits=(DML.ratios[order(DML.ratios$diff_H.L),]$chr_start)) + 
#   scale_y_continuous(limits=c(-80,80), breaks=c(-75, -50, -25, 0),
#                        labels=c("-75"="75%", "-50"="50%", "-25"="25%", "0"="0%"),
#                      sec.axis = sec_axis(trans=~.*1, breaks=c(75, 25, 50, 0),
#                        labels=c("-75"="75%", "-50"="50%", "-25"="25%", "0"="0%")))  +
#     theme(axis.text.y=element_blank(), axis.ticks.y = element_blank(), 
#           legend.position = "none", text = element_text(size=14, color="gray30"),
#           panel.background = element_blank(),
#           panel.border = element_rect(colour = "gray30", fill=NA, size=.4)) +
#     #ggtitle("DMLs, showing % methylation difference among  treatments") +
#     scale_fill_manual(values=c("firebrick3", "dodgerblue3")) +
#     #annotate(geom="text", x=-20, y=-65, label="Loci hypermethylated in\nSouth Sound", color="dodgerblue3") +
#     #annotate(geom="text", x=260, y=62, label="Loci hypermethylated in\nHood Canal", color="firebrick3") +
#   ylab("% methylation difference among  treatments") + xlab("Locus") + coord_flip()
# ggsave(filename = "../analyses/DMLs/DMLs-diff-barplot.pdf", width=7, height=5)
```




### Principal Component Analyses, DMLs only 

```{r, fig.show='hide'}
# load("../analyses/DMLs/R-objects/dml25_counts")
# 
# PCA.DMLs <- PCASamples(dml25_counts, obj.return = TRUE, sd.filter = T) #Run a PCA analysis on percent methylation for all samples. methylKit uses prcomp to create the PCA matrix
# nrow(PCA.DMLs$rotation)
# 
# summary(PCA.DMLs) #Look at summary statistics. The first PC explains 36.3% of variation, the second PC explains 8.5% of variation.  # new analysis/filtering AND >5x replace with NA: PC1=36.5% PC2=9.3%

```

### Save DML PC scores and R object for integration with other data 

```{r}
# save(PCA.DMLs, file="../analyses/DMLs/R-objects/PCA.DMLs")
# write_delim(data.frame(PCA.DMLs$x) %>% tibble::rownames_to_column("sample"), "../analyses/DMLs/PC-scores-DMLs.tab",  delim = '\t', col_names = TRUE)
```

### PCA Figure, DMLs only 

```{r}
# par(mar = c(5, 5, 1, 1)) #Specify inner and outer margins
# PCA.figure <- ordiplot(PCA.DMLs, choices = c(1, 2), type = "none", display = "sites", cex = 0.5, xlab = "", ylab = "", xaxt = "n", yaxt = "n") #Use ordiplot to create base biplot. Do not add any points
# points(PCA.figure, "sites", col = as.character(plotCustomization$color), pch = plotCustomization$symbol, cex = 2.5) #Add each sample. Darker samples are ambient, lighter samples are elevated pCO2
# #Add multiple white boxes on top of the default black box to manually change the color
# box(col = "white")
# box(col = "white")
# box(col = "white")
# box(col = "white")
# box(col = "white")
# box(col = "white")
# box(col = "white")
# box(col = "white")
# box(col = "white")
# box(col = "white")
# #ordiellipse(PCA.DMLs, plotCustomization$ treatment, show.groups = "HC", col = "firebrick3") #Add confidence ellipse around the samples in elevated pCO2
# #ordiellipse(PCA.DMLs, plotCustomization$ treatment, show.groups = "SS", col = "dodgerblue3") #Add confidence ellipse around the samples in ambient pCO2
# axis(side =  1, labels = TRUE, col = "grey80", cex.axis = 1.7) #Add x-axis
# mtext(side = 1, text = "PC 1 (36.5%)", line = 3, cex = 1.5) #Add x-axis label
# axis(side =  2, labels = TRUE, col = "grey80", cex.axis = 1.7) #Add y-axis
# mtext(side = 2, text = "PC 2 (9.3%)", line = 3, cex = 1.5) #Add y-axis label
# legend("topright", 
#        pch = c(16, 17), 
#        legend = c("Hood Canal", "South Sound"), 
#        col = c(c("firebrick3", "dodgerblue3")), 
#        cex = 1.4, bty = "n") #Add a legend with information about ambient and elevated samples
# 
# save(PCA.figure, file="../analyses/DMLs/R-objects/PCA.figure")
```

OTHER POSSIBLE ANALYSIS OPTIONS
- randomize which samples are "Treatment" and run 10 permutations to see now many loci are DMLs then 
- Try to pull SNPs from RNASeq data to identify pop structure 
- Research cool software for integrating meth+rnaseq. AI options for integration? 
